
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>macsypy.search_systems &#8212; MacSyFinder 1.0.4 documentation</title>
    <link rel="stylesheet" href="../../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '1.0.4',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">MacSyFinder documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../index.html" accesskey="U">Module code</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for macsypy.search_systems</h1><div class="highlight"><pre>
<span></span><span class="c1"># -*- coding: utf-8 -*-</span>

<span class="c1">################################################################################</span>
<span class="c1"># MacSyFinder - Detection of macromolecular systems in protein datasets        #</span>
<span class="c1">#               using systems modelling and similarity search.                 #</span>
<span class="c1"># Authors: Sophie Abby, Bertrand Néron                                         #</span>
<span class="c1"># Copyright © 2014  Institut Pasteur (Paris) and CNRS.                         #</span>
<span class="c1"># See the COPYRIGHT file for details                                           #</span>
<span class="c1">#                                                                              #</span>
<span class="c1"># MacsyFinder is distributed under the terms of the GNU General Public License #</span>
<span class="c1"># (GPLv3). See the COPYING file for details.                                   #</span>
<span class="c1">################################################################################</span>



<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">abc</span>
<span class="kn">import</span> <span class="nn">os.path</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">Counter</span><span class="p">,</span> <span class="n">OrderedDict</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">operator</span>
<span class="kn">import</span> <span class="nn">json</span>

<span class="kn">from</span> <span class="nn">macsypy_error</span> <span class="k">import</span> <span class="n">MacsypyError</span><span class="p">,</span> <span class="n">SystemDetectionError</span>
<span class="kn">from</span> <span class="nn">database</span> <span class="k">import</span> <span class="n">RepliconDB</span>
<span class="kn">from</span> <span class="nn">system</span> <span class="k">import</span> <span class="n">system_bank</span>

<span class="n">_log</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s1">&#39;macsyfinder.&#39;</span> <span class="o">+</span> <span class="vm">__name__</span><span class="p">)</span>

<span class="n">_log_out</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s1">&#39;macsyfinder.out&#39;</span><span class="p">)</span>

<div class="viewcode-block" id="ClustersHandler"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.ClustersHandler">[docs]</a><span class="k">class</span> <span class="nc">ClustersHandler</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Deals with sets of clusters found in a dataset. Conceived to store only clusters from a same replicon.</span>
<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="ClustersHandler.__init__"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.ClustersHandler.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param cfg: The configuration object built from default and user parameters.</span>
<span class="sd">        :type cfg: :class:`macsypy.config.Config` </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">replicon_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span></div>

    <span class="k">def</span> <span class="nf">add</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cluster</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">replicon_name</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">replicon_name</span> <span class="o">=</span> <span class="n">cluster</span><span class="o">.</span><span class="n">replicon_name</span>
            <span class="n">cluster</span><span class="o">.</span><span class="n">save</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cluster</span><span class="p">)</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">replicon_name</span> <span class="o">==</span> <span class="n">cluster</span><span class="o">.</span><span class="n">replicon_name</span><span class="p">:</span>
            <span class="n">cluster</span><span class="o">.</span><span class="n">save</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cluster</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;Attempting to add a cluster in a ClustersHandler dedicated to another replicon !&quot;</span>
            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">:</span>
                <span class="n">msg</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span>
            <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;To add: </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cluster</span><span class="p">)</span>
            <span class="n">_log</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">SystemDetectionError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">to_print</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">:</span>
            <span class="n">to_print</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="n">cluster</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">to_print</span>

<div class="viewcode-block" id="ClustersHandler.circularize"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.ClustersHandler.circularize">[docs]</a>    <span class="k">def</span> <span class="nf">circularize</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rep_info</span><span class="p">,</span> <span class="n">end_hits</span><span class="p">,</span> <span class="n">systems_to_detect</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This function takes into account the circularity of the replicon by merging clusters when appropriate</span>
<span class="sd">        (typically at replicon&#39;s ends).</span>
<span class="sd">        It has to be called only if the replicon_topology is set to \&quot;circular\&quot;.</span>

<span class="sd">        :param rep_info: an entry extracted from the :class:`macsypy.database.RepliconDB`</span>
<span class="sd">        :type rep_info: a namedTuple &quot;RepliconInfo&quot; :class:`macsypy.database.RepliconInfo`</span>
<span class="sd">        :param end_hits: a set of hits at ends of the replicon that were not introduced in clusters,</span>
<span class="sd">        and that might be part of a system overlapping the two &quot;ends&quot; of the replicon</span>
<span class="sd">        :type end_hits: a list of :class:`macsypy.report.Hit`</span>
<span class="sd">        :param systems_to_detect: the set of systems to detect in this run</span>
<span class="sd">        :type systems_to_detect: a list of :class:`macsypy.system.System</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># We assume this function is called when appropriate (i.e. for circular replicons)</span>

        <span class="n">pos_min</span> <span class="o">=</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">min</span>
        <span class="n">pos_max</span> <span class="o">=</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">max</span>

        <span class="n">check_clust</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">----------------------------------------</span><span class="se">\n</span><span class="s2">--- Handling of replicon circularity ---</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c1"># Here they might be the same !</span>
            <span class="n">clust_first</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">clust_last</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>

            <span class="c1"># check_clust = True</span>
            <span class="c1"># Case 1: none of the two end hits is part of a cluster =&gt; they might form a new cluster together. </span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">end_hits</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
                <span class="n">first</span> <span class="o">=</span> <span class="n">end_hits</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">second</span> <span class="o">=</span> <span class="n">end_hits</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="c1"># if(first.gene.position &gt; second.gene.position): # Genes do not have positions ! But Hits do!</span>
                <span class="k">if</span><span class="p">(</span><span class="n">first</span><span class="o">.</span><span class="n">position</span> <span class="o">&gt;</span> <span class="n">second</span><span class="o">.</span><span class="n">position</span><span class="p">):</span>
                    <span class="n">tmp</span> <span class="o">=</span> <span class="n">first</span>
                    <span class="n">first</span> <span class="o">=</span> <span class="n">second</span>
                    <span class="c1"># second = first</span>
                    <span class="n">second</span> <span class="o">=</span> <span class="n">tmp</span>
                <span class="n">dist</span> <span class="o">=</span> <span class="n">second</span><span class="o">.</span><span class="n">position</span> <span class="o">-</span> <span class="n">pos_min</span> <span class="o">+</span> <span class="n">pos_max</span> <span class="o">-</span> <span class="n">first</span><span class="o">.</span><span class="n">position</span> 
                <span class="k">if</span><span class="p">(</span><span class="n">dist</span> <span class="o">&lt;=</span> <span class="nb">max</span><span class="p">(</span><span class="n">first</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">inter_gene_max_space</span><span class="p">,</span> <span class="n">second</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">inter_gene_max_space</span><span class="p">)):</span>
                    <span class="c1"># If true, this means that there is no need to circularize other clusters as this cluster already</span>
                    <span class="c1"># overlaps the &quot;Ori&quot;. =&gt; check_clust is thus set to False</span>

                    <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;--- Two hits at both ends of the replicon form a new cluster.</span><span class="se">\n</span><span class="s2">&quot;</span>

                    <span class="n">new_clust</span> <span class="o">=</span> <span class="n">Cluster</span><span class="p">(</span><span class="n">systems_to_detect</span><span class="p">)</span>
                    <span class="n">new_clust</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">second</span><span class="p">)</span>
                    <span class="n">new_clust</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">first</span><span class="p">)</span>
                    <span class="n">new_clust</span><span class="o">.</span><span class="n">save</span><span class="p">()</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_clust</span><span class="p">)</span>
                    <span class="n">check_clust</span> <span class="o">=</span> <span class="kc">False</span>

            <span class="k">if</span> <span class="n">check_clust</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">end_hits</span><span class="p">:</span>
                    <span class="c1"># Two cases: we are at one end, or the other</span>
                    <span class="k">if</span> <span class="n">h</span><span class="o">.</span><span class="n">position</span> <span class="o">&gt;</span> <span class="n">clust_first</span><span class="o">.</span><span class="n">hits</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">position</span><span class="p">:</span>
                        <span class="c1"># Case 2: The end hit is at the terminal end of the replicon.</span>
                        <span class="c1"># Try to cluster it with the 1st stored cluster.</span>
                        <span class="c1"># print &quot;The end hit is at the TERMINAL end of the replicon. Test if it must be clustered with the 1st stored cluster.&quot;</span>
                        <span class="n">dist_end_hit</span> <span class="o">=</span> <span class="n">clust_first</span><span class="o">.</span><span class="n">begin</span> <span class="o">-</span> <span class="n">pos_min</span> <span class="o">+</span> <span class="n">pos_max</span> <span class="o">-</span> <span class="n">h</span><span class="o">.</span><span class="n">position</span> <span class="c1"># OK?</span>
                        <span class="k">if</span><span class="p">(</span><span class="n">dist_end_hit</span> <span class="o">&lt;=</span> <span class="nb">max</span><span class="p">(</span><span class="n">clust_first</span><span class="o">.</span><span class="n">hits</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">inter_gene_max_space</span><span class="p">,</span> <span class="n">h</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">inter_gene_max_space</span><span class="p">)):</span>
                            <span class="n">check_clust</span> <span class="o">=</span> <span class="kc">False</span>
                            <span class="c1">#print &quot;Cluster them !&quot;</span>

                            <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;--- A hit is at the TERMINAL end of the replicon, and must be clustered with the 1st stored cluster.</span><span class="se">\n</span><span class="s2">&quot;</span>

                            <span class="n">new_clust</span> <span class="o">=</span> <span class="n">Cluster</span><span class="p">(</span><span class="n">systems_to_detect</span><span class="p">)</span>
                            <span class="n">new_clust</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>
                            <span class="k">for</span> <span class="n">hit_clust</span> <span class="ow">in</span> <span class="n">clust_first</span><span class="o">.</span><span class="n">hits</span><span class="p">:</span>
                                <span class="n">new_clust</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">hit_clust</span><span class="p">)</span>

                            <span class="n">new_clust</span><span class="o">.</span><span class="n">save</span><span class="p">()</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_clust</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># Case 3: The end hit is at the initial end of the replicon.</span>
                        <span class="c1"># Try to cluster it with the last stored cluster.</span>
                        <span class="c1"># print &quot;The end hit is at the INITIAL end of the replicon. Test if it must be clustered with the last stored cluster.&quot;</span>
                        <span class="n">dist_end_hit</span> <span class="o">=</span> <span class="n">h</span><span class="o">.</span><span class="n">position</span> <span class="o">-</span> <span class="n">pos_min</span> <span class="o">+</span> <span class="n">pos_max</span> <span class="o">-</span> <span class="n">clust_last</span><span class="o">.</span><span class="n">end</span> 
                        <span class="k">if</span> <span class="n">dist_end_hit</span> <span class="o">&lt;=</span> <span class="nb">max</span><span class="p">(</span><span class="n">clust_last</span><span class="o">.</span><span class="n">hits</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">clust_last</span><span class="o">.</span><span class="n">hits</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">inter_gene_max_space</span><span class="p">,</span> <span class="n">h</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">inter_gene_max_space</span><span class="p">):</span>
                            <span class="n">check_clust</span> <span class="o">=</span> <span class="kc">False</span>
                            <span class="c1"># print &quot;Cluster them !&quot;</span>

                            <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;--- A hit is at the INITIAL end of the replicon, and must be clustered with the last stored cluster.</span><span class="se">\n</span><span class="s2">&quot;</span>

                            <span class="n">clust_last</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>
                            <span class="n">clust_last</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># Force to re-save the updated cluster</span>

        <span class="k">if</span> <span class="n">check_clust</span> <span class="ow">and</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">):</span>
            <span class="n">clust_first</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">clust_last</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

            <span class="c1"># Case 4: when putative end hits have been dealt with,</span>
            <span class="c1"># now remain cases where clusters are at both ends, and might be fused.</span>
            <span class="n">dist_clust</span> <span class="o">=</span> <span class="n">clust_first</span><span class="o">.</span><span class="n">begin</span> <span class="o">-</span> <span class="n">pos_min</span> <span class="o">+</span> <span class="n">pos_max</span> <span class="o">-</span> <span class="n">clust_last</span><span class="o">.</span><span class="n">end</span>

            <span class="k">if</span> <span class="n">dist_clust</span> <span class="o">&lt;=</span> <span class="nb">max</span><span class="p">(</span><span class="n">clust_first</span><span class="o">.</span><span class="n">hits</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">inter_gene_max_space</span><span class="p">,</span> <span class="n">clust_last</span><span class="o">.</span><span class="n">hits</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">clust_last</span><span class="o">.</span><span class="n">hits</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">inter_gene_max_space</span><span class="p">):</span>
                <span class="c1"># Need to circularize !</span>
                <span class="c1">#print &quot;A cluster needs to be \&quot;circularized\&quot; ! &quot;</span>
                <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;--- Two clusters should be merged into a new cluster </span><span class="se">\&quot;</span><span class="s2">circularized</span><span class="se">\&quot;</span><span class="s2"> !</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="c1"># The &quot;1st&quot; cluster on the replicon is removed,</span>
                <span class="c1"># as its hits will be appended to the &quot;last&quot; cluster on the replicon.</span>
                <span class="c1">#self.clusters.pop(0) </span>
                <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">clust_first</span><span class="o">.</span><span class="n">hits</span><span class="p">:</span>
                    <span class="n">clust_last</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span> 
                <span class="n">clust_last</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># Force to re-save the updated cluster</span>
                <span class="c1">#print clust_last</span>
                <span class="n">msg</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="n">clust_last</span><span class="p">)</span>

        <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;----------------------------------------&quot;</span>
        <span class="n">_log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span></div></div>



<div class="viewcode-block" id="Cluster"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.Cluster">[docs]</a><span class="k">class</span> <span class="nc">Cluster</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Stores a set of contiguous hits. The Cluster object can have different states regarding </span>
<span class="sd">    its content in different genes&#39; systems: </span>

<span class="sd">      - ineligible: not a cluster to analyze</span>
<span class="sd">      - clear: a single system is represented in the cluster</span>
<span class="sd">      - ambiguous: several systems are represented in the cluster =&gt; might need a disambiguation</span>

<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="Cluster.__init__"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.Cluster.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">systems_to_detect</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param systems_to_detect: the list of systems to be detected in this run</span>
<span class="sd">        :type systems_to_detect: a list of :class:`macsypy.system.System`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">hits</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">systems_to_detect</span> <span class="o">=</span> <span class="n">systems_to_detect</span> <span class="c1"># NEW</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">systems</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">replicon_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">begin</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_state</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_putative_system</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_compatible_systems</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1"># NEW!</span></div>

<div class="viewcode-block" id="Cluster.__len__"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.Cluster.__len__">[docs]</a>    <span class="k">def</span> <span class="nf">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :return: the length of the Cluster, *i.e.*, the number of hits stored in it</span>
<span class="sd">        :rtype: integer</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">hits</span><span class="p">)</span></div>

<div class="viewcode-block" id="Cluster.__str__"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.Cluster.__str__">[docs]</a>    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        print of the Cluster&#39;s hits stored in terms of components, and corresponding sequence identifier and positions</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">pos</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">seq_ids</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">gene_names</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hits</span><span class="p">:</span>
            <span class="n">pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">h</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>
            <span class="n">seq_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">h</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
            <span class="n">gene_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">h</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="n">clear</span> <span class="o">=</span> <span class="s1">&#39;?&#39;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">state</span> <span class="o">==</span> <span class="s2">&quot;clear&quot;</span><span class="p">:</span>
            <span class="n">clear</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
        <span class="k">return</span> <span class="s2">&quot;--- Cluster </span><span class="si">{0}</span><span class="s2"> </span><span class="si">{1}</span><span class="s2"> ---</span><span class="se">\n</span><span class="si">{2}</span><span class="se">\n</span><span class="si">{3}</span><span class="se">\n</span><span class="si">{4}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">putative_system</span><span class="p">,</span>
                                                               <span class="n">clear</span><span class="p">,</span> 
                                                               <span class="n">seq_ids</span><span class="p">,</span> 
                                                               <span class="n">gene_names</span><span class="p">,</span> 
                                                               <span class="n">pos</span><span class="p">)</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">state</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :return: the state of the Cluster of hits</span>
<span class="sd">        :rtype: string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_state</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">putative_system</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :return: the name of the putative system represented by the cluster</span>
<span class="sd">        :rtype: string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_putative_system</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">compatible_systems</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :return: the list of the names of compatible systems represented by the cluster</span>
<span class="sd">        :rtype: string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_compatible_systems</span>

<div class="viewcode-block" id="Cluster.add"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.Cluster.add">[docs]</a>    <span class="k">def</span> <span class="nf">add</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">hit</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add a Hit to a Cluster.</span>
<span class="sd">        Hits are always added at the end of the cluster (appended to the list of hits).</span>
<span class="sd">        Thus, &#39;begin&#39; and &#39;end&#39; positions of the Cluster are always the position of the 1st and of the last hit respectively.</span>

<span class="sd">        :param hit: the Hit to add</span>
<span class="sd">        :type hit: a :class:`macsypy.report.Hit`</span>
<span class="sd">        :raise: a :class:`macsypy.macsypy_error.SystemDetectionError`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># need to update cluster bounds</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">hits</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">begin</span> <span class="o">=</span> <span class="n">hit</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">begin</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">replicon_name</span> <span class="o">=</span> <span class="n">hit</span><span class="o">.</span><span class="n">replicon_name</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">hit</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">replicon_name</span> <span class="o">==</span> <span class="n">hit</span><span class="o">.</span><span class="n">replicon_name</span><span class="p">:</span>
                <span class="c1"># To be updated !! make this work also with &quot;circularized&quot; clusters</span>
                <span class="c1"># The end position only is updated, as Hits are always appended.</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">=</span> <span class="n">hit</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">hit</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;Attempting to gather in a cluster hits from different replicons ! &quot;</span>
                <span class="n">_log</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
                <span class="k">raise</span> <span class="n">SystemDetectionError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span></div>


<div class="viewcode-block" id="Cluster.save"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.Cluster.save">[docs]</a>    <span class="k">def</span> <span class="nf">save</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Check the status of the cluster regarding systems which have hits in it. </span>
<span class="sd">        Update systems represented, and assign a putative system (self._putative_system),</span>
<span class="sd">        which is the system with most hits in the cluster.</span>
<span class="sd">        The systems represented are stored in a dictionary in the self.systems variable. </span>
<span class="sd">        The execution of this function can be forced, even if it has already run</span>
<span class="sd">        for the cluster with the option force=True.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">putative_system</span> <span class="ow">or</span> <span class="n">force</span><span class="p">:</span>
            <span class="c1"># First compute the &quot;Majoritary&quot; system</span>
            <span class="n">systems</span> <span class="o">=</span> <span class="p">{}</span> <span class="c1"># Counter of occcurrences of systems in the cluster (keys are systems&#39; names)</span>
            <span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="c1">#systems_object={} # Store systems object. # To be replaced by &quot;system_bank&quot;? Yep ! done</span>

            <span class="c1"># Version with hits &quot;reference&quot; systems</span>
            <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hits</span><span class="p">:</span>
                <span class="n">syst</span> <span class="o">=</span> <span class="n">h</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">name</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">systems</span><span class="o">.</span><span class="n">has_key</span><span class="p">(</span><span class="n">syst</span><span class="p">):</span>
                    <span class="n">systems</span><span class="p">[</span><span class="n">syst</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="c1">#systems_object[syst]=h.system</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">systems</span><span class="p">[</span><span class="n">syst</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="n">genes</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">h</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">h</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">systems</span> <span class="o">=</span> <span class="n">systems</span>

            <span class="c1"># Useless ?! Or change? </span>
            <span class="n">max_syst</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">tmp_syst_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">systems</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">y</span> <span class="o">&gt;=</span> <span class="n">max_syst</span><span class="p">:</span>
                    <span class="n">tmp_syst_name</span> <span class="o">=</span> <span class="n">x</span>
                    <span class="n">max_syst</span> <span class="o">=</span> <span class="n">y</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_putative_system</span> <span class="o">=</span> <span class="n">tmp_syst_name</span> <span class="c1"># Remove cause useless?</span>

            <span class="c1"># NEW Version with hits all &quot;compatible&quot; systems</span>
            <span class="n">systems_compat</span> <span class="o">=</span> <span class="p">{}</span> <span class="c1"># Counter of occcurrences of COMPATIBLE (-extended- list of) systems in the cluster. Keys are systems&#39; names</span>
            <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hits</span><span class="p">:</span>
                <span class="c1">#syst_list=h.gene.get_compatible_systems(self.systems_to_detect) # Need the list of systems (obj!) to be detected... in the cfg?</span>
                <span class="c1"># Now exclude forbidden genes from those that define the list of compatible systems</span>
                <span class="n">syst_list</span> <span class="o">=</span> <span class="n">h</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">get_compatible_systems</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">systems_to_detect</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># Need the list of systems (obj!) to be detected... in the cfg? # tmp before nope</span>
                <span class="c1">#syst_list=h.gene.get_compatible_systems(self.systems_to_detect, True) # Need the list of systems (obj!) to be detected... in the cfg? # tmp before yep</span>
                <span class="k">for</span> <span class="n">syst</span> <span class="ow">in</span> <span class="n">syst_list</span><span class="p">:</span>
                    <span class="n">syst_name</span> <span class="o">=</span> <span class="n">syst</span><span class="o">.</span><span class="n">name</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">systems_compat</span><span class="o">.</span><span class="n">has_key</span><span class="p">(</span><span class="n">syst_name</span><span class="p">):</span>
                        <span class="n">systems_compat</span><span class="p">[</span><span class="n">syst_name</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                        <span class="c1">#systems_object[syst]=h.system</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">systems_compat</span><span class="p">[</span><span class="n">syst_name</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="n">genes</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">h</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">h</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>

            <span class="c1">#print systems_compat</span>
            <span class="c1"># We sort the list of compatible systems per decreasing nb of systems occurrences</span>
            <span class="n">systems_compat</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">systems_compat</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">t</span><span class="p">:</span> <span class="n">t</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
            <span class="c1">#print systems_compat</span>
            <span class="c1">#print self</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">hits</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">loner</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_state</span> <span class="o">=</span> <span class="s2">&quot;ineligible&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Check for foreign &quot;accessory&quot; genes... Might increase nb of systems predicted in the cluster,</span>
                <span class="c1"># even if they are tolerated in the cluster.</span>
                <span class="c1"># Also deal with foreign &quot;exchangeable&quot; genes for the same reasons... NB !!</span>
                <span class="c1"># Maybe just not add the system to the list if exchangeable?</span>
                <span class="c1">#if len(systems.keys()) == 1:</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">systems_compat</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_state</span> <span class="o">=</span> <span class="s2">&quot;clear&quot;</span>
                    <span class="c1">#syst = systems.keys()[0]</span>
                    <span class="n">syst</span> <span class="o">=</span> <span class="n">systems_compat</span><span class="o">.</span><span class="n">keys</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="c1">#print syst</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_putative_system</span> <span class="o">=</span> <span class="n">syst</span>
                    <span class="c1">#self._compatible_systems.append(system_bank[syst])</span>
                    <span class="c1"># Store only compatible systems that are searched for !!</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_compatible_systems</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">syst</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Check for foreign &quot;accessory&quot; genes regarding the majoritary system...</span>
                    <span class="c1"># They might increase nb of systems predicted in the cluster artificially,</span>
                    <span class="c1"># even if they are tolerated in the cluster.</span>
                    <span class="c1"># For that need to scan again all hits and ask wether they are accessory foreign genes.</span>
                    <span class="k">def</span> <span class="nf">try_system</span><span class="p">(</span><span class="n">hits</span><span class="p">,</span> <span class="n">putative_system</span><span class="p">,</span> <span class="n">counter_systems_in_clust</span><span class="p">):</span>
                        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                        Test if the putative_system is compatible with the systems of hits (counter_systems_in_clust)</span>

<span class="sd">                        :param hits: a list of hits</span>
<span class="sd">                        :type hits: a list of :class:`macsypy.report.Hit`</span>
<span class="sd">                        :param putative_system: the name of a putative system to consider</span>
<span class="sd">                        :type putative_system: string</span>
<span class="sd">                        :param counter_systems_in_clust: a dictionary with systems occurrences when exploring the hits</span>
<span class="sd">                        :type counter_systems_in_clust: a dictionary with systems&#39; names as keys, and counts as values</span>
<span class="sd">                        :return: the &quot;state&quot; of the set of hits regarding the putative system</span>
<span class="sd">                        :rtype: string</span>

<span class="sd">                        &quot;&quot;&quot;</span>
                        <span class="c1">#foreign_accessory = 0</span>
                        <span class="n">auth</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># counts nb of hits that are authorized in the putative system</span>
                        <span class="c1">#forbid = 0</span>
                        <span class="c1">#print &quot;Unclear state with multiple systems to deal with...&quot;</span>
                        <span class="c1">#print systems</span>
                        <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">hits</span><span class="p">:</span>
                            <span class="c1"># Exclude the consideration of &quot;forbidden&quot; genes !</span>
                            <span class="c1">#if h.gene.is_authorized(system_bank[putative_system], False): # tmp before nope</span>
                            <span class="k">if</span> <span class="n">h</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">is_authorized</span><span class="p">(</span><span class="n">system_bank</span><span class="p">[</span><span class="n">putative_system</span><span class="p">],</span> <span class="kc">True</span><span class="p">):</span> <span class="c1"># No! forbidden genes that are defined in system has to be considered!</span>
                                <span class="n">auth</span> <span class="o">+=</span> <span class="mi">1</span>
                            <span class="c1">#if h.gene.is_forbidden(system_bank[putative_system]):</span>
                            <span class="c1">#    forbid +=1</span>
                                                    
                            <span class="k">if</span> <span class="n">auth</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">hits</span><span class="p">):</span>
                                <span class="n">state</span> <span class="o">=</span> <span class="s2">&quot;clear&quot;</span>
                                <span class="c1">#print &quot;clear {0}&quot;.format(putative_system)</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">state</span> <span class="o">=</span> <span class="s2">&quot;ambiguous&quot;</span>
                                <span class="c1">#print &quot;ambiguous {0}&quot;.format(putative_system)</span>

<span class="c1">#                         if forbid == 0:</span>
<span class="c1">#                             if auth == len(hits):</span>
<span class="c1">#                                 state = &quot;clear&quot;</span>
<span class="c1">#                                 #print &quot;clear {}&quot;.format(putative_system)</span>
<span class="c1">#                             else:</span>
<span class="c1">#                                 state = &quot;ambiguous&quot;</span>
<span class="c1">#                                 #print &quot;ambiguous {}&quot;.format(putative_system)</span>
<span class="c1">#                         else:</span>
<span class="c1">#                             state = &quot;ineligible&quot;</span>
<span class="c1">#                             #print &quot;ineligible {}&quot;.format(putative_system)</span>
                        <span class="k">return</span> <span class="n">state</span>

                    <span class="c1"># Sort systems to consider by decreasing counts.</span>
                    <span class="n">cluster_compatible_systems</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="k">for</span> <span class="n">putative_system</span> <span class="ow">in</span> <span class="n">systems_compat</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                        <span class="c1"># Add that it has to be done first from the most rep systems by decreasing order of systems.values.</span>
                        <span class="n">state</span> <span class="o">=</span> <span class="n">try_system</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">hits</span><span class="p">,</span> <span class="n">putative_system</span><span class="p">,</span> <span class="n">systems_compat</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">state</span> <span class="o">==</span> <span class="s2">&quot;clear&quot;</span><span class="p">:</span>
                            <span class="c1">#print &quot;BUENO SYSTEMO {0}&quot;.format(putative_system)</span>
                            <span class="c1">#self._state=&quot;clear&quot; </span>
                            <span class="c1">#self._putative_system=putative_system </span>
                            <span class="c1">#break</span>
                            <span class="n">cluster_compatible_systems</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">putative_system</span><span class="p">)</span>
                        <span class="c1">#else:</span>
                        <span class="c1">#    self._state=&quot;ambiguous&quot; # tmp before nope</span>
                        <span class="c1">#    # Aoutch in this case no putative_system?!</span>
                        <span class="c1">#    print &quot;YAPABON...{0}&quot;.format(putative_system)</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cluster_compatible_systems</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="c1">#print cluster_compatible_systems</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_state</span> <span class="o">=</span> <span class="s2">&quot;clear&quot;</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_putative_system</span> <span class="o">=</span> <span class="n">cluster_compatible_systems</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_compatible_systems</span> <span class="o">=</span> <span class="n">cluster_compatible_systems</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_state</span> <span class="o">=</span> <span class="s2">&quot;ambiguous&quot;</span></div></div>
        <span class="c1">#print self._compatible_systems</span>



<div class="viewcode-block" id="SystemNameGenerator"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemNameGenerator">[docs]</a><span class="k">class</span> <span class="nc">SystemNameGenerator</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates and stores the names of detected systems. Ensures the uniqueness of the names.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">name_bank</span> <span class="o">=</span> <span class="p">{}</span>

<div class="viewcode-block" id="SystemNameGenerator.getSystemName"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemNameGenerator.getSystemName">[docs]</a>    <span class="k">def</span> <span class="nf">getSystemName</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">replicon</span><span class="p">,</span> <span class="n">system</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates a unique system name based on the replicon&#39;s name and the system&#39;s name.</span>

<span class="sd">        :param replicon: the replicon name</span>
<span class="sd">        :type replicon: string</span>
<span class="sd">        :param system: the system name</span>
<span class="sd">        :type system: string</span>
<span class="sd">        :return: a unique system name</span>
<span class="sd">        :rtype: string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">basename</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_computeBasename</span><span class="p">(</span><span class="n">replicon</span><span class="p">,</span> <span class="n">system</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">basename</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">name_bank</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">name_bank</span><span class="p">[</span><span class="n">basename</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">name_bank</span><span class="p">[</span><span class="n">basename</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="n">system_name</span> <span class="o">=</span> <span class="n">basename</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name_bank</span><span class="p">[</span><span class="n">basename</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">system_name</span></div>

<div class="viewcode-block" id="SystemNameGenerator._computeBasename"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemNameGenerator._computeBasename">[docs]</a>    <span class="k">def</span> <span class="nf">_computeBasename</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">replicon</span><span class="p">,</span> <span class="n">system</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Computes the base name to be used for unique name generation</span>

<span class="sd">        :param replicon: the replicon name</span>
<span class="sd">        :type replicon: string</span>
<span class="sd">        :param system: the system name</span>
<span class="sd">        :type system: string</span>
<span class="sd">        :return: the base name</span>
<span class="sd">        :rtype: string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{1}</span><span class="s2">_&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">replicon</span><span class="p">,</span> <span class="n">system</span><span class="p">)</span></div></div>

<span class="n">system_name_generator</span> <span class="o">=</span> <span class="n">SystemNameGenerator</span><span class="p">()</span>



<div class="viewcode-block" id="SystemOccurence"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence">[docs]</a><span class="k">class</span> <span class="nc">SystemOccurence</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This class is instantiated for a specific system that has been asked for detection. It can be filled step by step with hits. </span>
<span class="sd">    A decision can then be made according to the parameters defined *e.g.* quorum of genes. </span>

<span class="sd">    The SystemOccurence object has a &quot;state&quot; parameter, with the possible following values: </span>
<span class="sd">      - &quot;empty&quot; if the SystemOccurence has not yet been filled with genes of the decision rule of the system</span>
<span class="sd">      - &quot;no_decision&quot; if the filling process has started but the decision rule has not yet been applied to this occurence</span>
<span class="sd">      - &quot;single_locus&quot;</span>
<span class="sd">      - &quot;multi_loci&quot;</span>
<span class="sd">      - &quot;uncomplete&quot;</span>

<span class="sd">    &quot;&quot;&quot;</span>
<div class="viewcode-block" id="SystemOccurence.__init__"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">system</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param system: the system to \&quot;fill\&quot; with hits.</span>
<span class="sd">        :type system: :class:`macsypy.system.System` </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">system</span> <span class="o">=</span> <span class="n">system</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span> <span class="o">=</span> <span class="n">system</span><span class="o">.</span><span class="n">name</span>

        <span class="c1"># Variables to be updated during the system detection </span>
        <span class="bp">self</span><span class="o">.</span><span class="n">valid_hits</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1"># validSystemHit are stored with the &quot;fill_with&quot; function, and ready for extraction in case of a positive detection</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">loci_positions</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1"># list of tuples</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_state</span> <span class="o">=</span> <span class="s2">&quot;empty&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nb_cluster</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_nb_syst_genes</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">unique_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

        <span class="c1"># System definition</span>
        <span class="c1"># Make those attributes non modifiable?</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">exmandatory_genes</span> <span class="o">=</span> <span class="p">{}</span> <span class="c1"># List of &#39;exchanged&#39; mandatory genes</span>

        <span class="c1"># New ! Add of a list of &quot;multi_system&quot; genes, fed only from mandatory and accessory genes from the actual system (and not &#39;exchanged&#39;)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">multi_syst_genes</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">system</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">[</span><span class="n">g</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">exchangeable</span><span class="p">:</span>
                <span class="n">homologs</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">get_homologs</span><span class="p">()</span>
                <span class="n">analogs</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">get_analogs</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">ex</span> <span class="ow">in</span> <span class="n">homologs</span> <span class="o">+</span> <span class="n">analogs</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">exmandatory_genes</span><span class="p">[</span><span class="n">ex</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">name</span>
            <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">multi_system</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">multi_syst_genes</span><span class="p">[</span><span class="n">g</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">exaccessory_genes</span> <span class="o">=</span> <span class="p">{}</span> <span class="c1"># List of &#39;exchanged&#39; accessory genes</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">system</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">[</span><span class="n">g</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">exchangeable</span><span class="p">:</span>
                <span class="n">homologs</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">get_homologs</span><span class="p">()</span>
                <span class="n">analogs</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">get_analogs</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">ex</span> <span class="ow">in</span> <span class="n">homologs</span> <span class="o">+</span> <span class="n">analogs</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">exaccessory_genes</span><span class="p">[</span><span class="n">ex</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">name</span>
            <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">multi_system</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">multi_syst_genes</span><span class="p">[</span><span class="n">g</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">forbidden_genes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">exforbidden_genes</span> <span class="o">=</span> <span class="p">{}</span> <span class="c1"># List of &#39;exchanged&#39; forbidden genes</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">system</span><span class="o">.</span><span class="n">forbidden_genes</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">forbidden_genes</span><span class="p">[</span><span class="n">g</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">exchangeable</span><span class="p">:</span>
                <span class="n">homologs</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">get_homologs</span><span class="p">()</span>
                <span class="n">analogs</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">get_analogs</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">ex</span> <span class="ow">in</span> <span class="n">homologs</span> <span class="o">+</span> <span class="n">analogs</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">exforbidden_genes</span><span class="p">[</span><span class="n">ex</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">name</span></div>
            <span class="c1"># Forbidden genes do not play a role in the system, thus they do not have the multi_system feature</span>
            <span class="c1">#if g.multi_system:</span>
            <span class="c1">#    self.multi_syst_genes[g.name] = 0</span>


<div class="viewcode-block" id="SystemOccurence.get_gene_ref"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.get_gene_ref">[docs]</a>    <span class="k">def</span> <span class="nf">get_gene_ref</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param gene: the gene to get it&#39;s gene reference</span>
<span class="sd">        :type gene: :class:`macsypy.gene.Gene`, or :class:`macsypy.gene.Homolog` or :class:`macsypy.gene.Analog` object</span>
<span class="sd">        :return: object :class:`macsypy.gene.Gene` or None </span>
<span class="sd">        :rtype: :class:`macsypy.gene.Gene` object or None</span>
<span class="sd">        :raise: KeyError if the system does not contain any gene gene.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">get_gene_ref</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></div>


<div class="viewcode-block" id="SystemOccurence.__str__"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.__str__">[docs]</a>    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :return: Information of the component content of the SystemOccurence.</span>
<span class="sd">        :rtype: string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">:</span> 
            <span class="n">out</span> <span class="o">+=</span> <span class="s2">&quot;Mandatory genes: </span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
                <span class="n">out</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="se">\t</span><span class="si">{1:d}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">g</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">+=</span> <span class="s2">&quot;Accessory genes: </span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
                <span class="n">out</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="se">\t</span><span class="si">{1:d}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">g</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">forbidden_genes</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">+=</span> <span class="s2">&quot;Forbidden genes: </span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">forbidden_genes</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
                <span class="n">out</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="se">\t</span><span class="si">{1:d}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">g</span><span class="p">)</span>
        <span class="c1"># NEW</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">multi_syst_genes</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">+=</span> <span class="s2">&quot;Multi_syst genes:</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">multi_syst_genes</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
                <span class="n">out</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="se">\t</span><span class="si">{1:d}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">g</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">out</span></div>

<div class="viewcode-block" id="SystemOccurence.get_gene_counter_output"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.get_gene_counter_output">[docs]</a>    <span class="k">def</span> <span class="nf">get_gene_counter_output</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">forbid_exclude</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param forbid_exclude: exclude the forbidden components if set to True. False by default.</span>
<span class="sd">        :type forbid_exclude: boolean</span>
<span class="sd">        :returns: A dictionary ready for printing in system summary, \</span>
<span class="sd">         with genes (mandatory, accessory and forbidden if specified) occurences in the system occurrence.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">out</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">out</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">)</span>
        <span class="n">out</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">forbid_exclude</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">forbidden_genes</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{}</span><span class="s2">&quot;</span>
        <span class="k">return</span> <span class="n">out</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">state</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :return: the state of the systemOccurrence.</span>
<span class="sd">        :rtype: string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_state</span>

<div class="viewcode-block" id="SystemOccurence.get_system_unique_name"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.get_system_unique_name">[docs]</a>    <span class="k">def</span> <span class="nf">get_system_unique_name</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">replicon_name</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attributes unique name to the system occurrence with the class :class:`macsypy.search_systems.SystemNameGenerator`.</span>
<span class="sd">        Generates the name if not already set. </span>

<span class="sd">        :param replicon_name: the name of the replicon</span>
<span class="sd">        :type replicon_name: string</span>
<span class="sd">        :return: the unique name of the :class:`macsypy.search_systems.SystemOccurence`</span>
<span class="sd">        :rtype: string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">unique_name</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">unique_name</span> <span class="o">=</span> <span class="n">system_name_generator</span><span class="o">.</span><span class="n">getSystemName</span><span class="p">(</span><span class="n">replicon_name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">unique_name</span></div>

<div class="viewcode-block" id="SystemOccurence.get_system_name_unordered"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.get_system_name_unordered">[docs]</a>    <span class="k">def</span> <span class="nf">get_system_name_unordered</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;_putative&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attributes a name to the system occurrence for an &quot;unordered&quot; dataset =&gt; generating a generic name based</span>
<span class="sd">         on the system name and the suffix given.</span>

<span class="sd">        :param suffix: the suffix to be used for generating the systemOccurrence&#39;s name</span>
<span class="sd">        :type suffix: string</span>
<span class="sd">        :return: a name for a system in an &quot;unordered&quot; dataset to the :class:`macsypy.search_systems.SystemOccurence`</span>
<span class="sd">        :rtype: string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span> <span class="o">+</span> <span class="n">suffix</span></div>


<div class="viewcode-block" id="SystemOccurence.compute_system_length"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.compute_system_length">[docs]</a>    <span class="k">def</span> <span class="nf">compute_system_length</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rep_info</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the length of the system, all loci gathered, in terms of protein number</span>
<span class="sd">        (even those not matching any system gene)</span>

<span class="sd">        :param rep_info: an entry extracted from the :class:`macsypy.database.RepliconDB`</span>
<span class="sd">        :type rep_info: a namedTuple &quot;RepliconInfo&quot; :class:`macsypy.database.RepliconInfo`</span>
<span class="sd">        :rtype: integer</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">length</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="c1"># To be updated to deal with &quot;circular&quot; clusters</span>
        <span class="k">for</span><span class="p">(</span><span class="n">begin</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">loci_positions</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">begin</span> <span class="o">&lt;=</span> <span class="n">end</span><span class="p">:</span>
                <span class="n">length</span> <span class="o">+=</span> <span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">begin</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">topology</span> <span class="o">==</span> <span class="s2">&quot;circular&quot;</span><span class="p">:</span>
                <span class="n">locus_length</span> <span class="o">=</span> <span class="n">end</span> <span class="o">-</span> <span class="n">begin</span> <span class="o">+</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">max</span> <span class="o">-</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">min</span> <span class="o">+</span> <span class="mi">2</span>
                <span class="n">length</span> <span class="o">+=</span> <span class="n">locus_length</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;Inconsistency in locus positions in the case of a linear replicon.</span><span class="se">\</span>
<span class="s2"> The begin position of a locus cannot be higher than the end position. </span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;Problem with locus found with positions begin: </span><span class="si">{0:d}</span><span class="s2"> end: </span><span class="si">{1:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">begin</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
                <span class="n">_log</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
                <span class="k">raise</span> <span class="n">SystemDetectionError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">length</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">nb_syst_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This value is set after a decision was made on the system in</span>
<span class="sd">        :func:`macsypy.search_systems.SystemOccurence:decision_rule`</span>

<span class="sd">        :return: the number of mandatory and accessory genes with at least one occurence</span>
<span class="sd">        (number of different accessory genes)</span>
<span class="sd">        :rtype: integer</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_nb_syst_genes</span>

    <span class="k">def</span> <span class="nf">compute_nb_syst_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">)</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">compute_nb_syst_genes_tot</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_genes_tot</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">)</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_genes_tot</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">)</span>

<div class="viewcode-block" id="SystemOccurence.count_genes"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.count_genes">[docs]</a>    <span class="k">def</span> <span class="nf">count_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_dict</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Counts the number of genes with at least one occurrence in a dictionary with a counter of genes. </span>

<span class="sd">        :param gene_dict: a dictionary with gene&#39;s names as keys and number of occurrences as values</span>
<span class="sd">        :type gene_dict: dict</span>
<span class="sd">        :rtype: integer</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">total</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">gene_dict</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">v</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">total</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="n">total</span></div>

<div class="viewcode-block" id="SystemOccurence.count_genes_tot"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.count_genes_tot">[docs]</a>    <span class="k">def</span> <span class="nf">count_genes_tot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_dict</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Counts the number of matches in a dictionary with a counter of genes, independently of the nb of genes matched.</span>

<span class="sd">        :param gene_dict: a dictionary with gene&#39;s names as keys and number of occurrences as values</span>
<span class="sd">        :type gene_dict: dict</span>
<span class="sd">        :rtype: integer</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">total</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">gene_dict</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">total</span> <span class="o">+=</span> <span class="n">v</span>
        <span class="k">return</span> <span class="n">total</span></div>

<div class="viewcode-block" id="SystemOccurence.compute_missing_genes_list"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.compute_missing_genes_list">[docs]</a>    <span class="k">def</span> <span class="nf">compute_missing_genes_list</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_dict</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param gene_dict: a dictionary with gene&#39;s names as keys and number of occurrences as values</span>
<span class="sd">        :type gene_dict: dict</span>
<span class="sd">        :returns: the list of genes with no occurence in the gene counter. </span>
<span class="sd">        :rtype: list</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">missing</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">gene_dict</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">v</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">missing</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">missing</span></div>


<div class="viewcode-block" id="SystemOccurence.count_missing_genes"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.count_missing_genes">[docs]</a>    <span class="k">def</span> <span class="nf">count_missing_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_dict</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Counts the number of genes with no occurrence in the gene counter.</span>

<span class="sd">        :param gene_dict: a dictionary with gene&#39;s names as keys and number of occurrences as values</span>
<span class="sd">        :type gene_dict: dict</span>
<span class="sd">        :rtype: integer</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">compute_missing_genes_list</span><span class="p">(</span><span class="n">gene_dict</span><span class="p">))</span></div>


<div class="viewcode-block" id="SystemOccurence.is_complete"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.is_complete">[docs]</a>    <span class="k">def</span> <span class="nf">is_complete</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Test for SystemOccurrence completeness.</span>

<span class="sd">        :returns: True if the state of the SystemOccurrence is &quot;single_locus&quot; or &quot;multi_loci&quot;, False otherwise.</span>
<span class="sd">        :rtype: boolean</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">state</span> <span class="o">==</span> <span class="s2">&quot;single_locus&quot;</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">state</span> <span class="o">==</span> <span class="s2">&quot;multi_loci&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span> </div>

<div class="viewcode-block" id="SystemOccurence.get_summary_header"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.get_summary_header">[docs]</a>    <span class="k">def</span> <span class="nf">get_summary_header</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a string with the description of the summary returned by self.get_summary()</span>

<span class="sd">        :rtype: string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s2">&quot;#Replicon_name</span><span class="se">\t</span><span class="s2">System_Id</span><span class="se">\t</span><span class="s2">Reference_system</span><span class="se">\t</span><span class="s2">System_status</span><span class="se">\t</span><span class="s2">Nb_loci</span><span class="se">\t</span><span class="s2">Nb_Ref_mandatory</span><span class="se">\t</span><span class="s2">Nb_Ref_accessory</span><span class="se">\</span>
<span class="se">\t</span><span class="s2">Nb_Ref_Genes_detected_NR</span><span class="se">\t</span><span class="s2">Nb_Genes_with_match</span><span class="se">\t</span><span class="s2">System_length</span><span class="se">\t</span><span class="s2">Nb_Mandatory_NR</span><span class="se">\t</span><span class="s2">Nb_Accessory_NR</span><span class="se">\</span>
<span class="se">\t</span><span class="s2">Nb_missing_mandatory</span><span class="se">\t</span><span class="s2">Nb_missing_accessory</span><span class="se">\t</span><span class="s2">List_missing_mandatory</span><span class="se">\t</span><span class="s2">List_missing_accessory</span><span class="se">\t</span><span class="s2">Loci_positions</span><span class="se">\</span>
<span class="se">\t</span><span class="s2">Occur_Mandatory</span><span class="se">\t</span><span class="s2">Occur_Accessory</span><span class="se">\t</span><span class="s2">Occur_Forbidden&quot;</span></div>


<div class="viewcode-block" id="SystemOccurence.get_summary"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.get_summary">[docs]</a>    <span class="k">def</span> <span class="nf">get_summary</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">replicon_name</span><span class="p">,</span> <span class="n">rep_info</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Gives a summary of the system occurrence in terms of gene content and localization.</span>

<span class="sd">        :param replicon_name: the name of the replicon</span>
<span class="sd">        :type replicon_name: string</span>
<span class="sd">        :param rep_info: an entry extracted from the :class:`macsypy.database.RepliconDB`</span>
<span class="sd">        :type rep_info: a namedTuple &quot;RepliconInfo&quot; :class:`macsypy.database.RepliconInfo`        </span>
<span class="sd">        :return: a tabulated summary of the :class:`macsypy.search_systems.SystemOccurence`</span>
<span class="sd">        :rtype: string</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">report_str</span> <span class="o">=</span> <span class="n">replicon_name</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_system_unique_name</span><span class="p">(</span><span class="n">replicon_name</span><span class="p">)</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">)</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">state</span><span class="p">)</span>
        <span class="c1"># Nb of loci included to fill the system occurrence</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nb_cluster</span><span class="p">)</span>
        <span class="c1"># Nb mandatory_genes in the definition of the system</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">))</span>
        <span class="c1"># Nb accessory_genes in the definition of the system</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">))</span>
        <span class="c1"># Nb syst genes NR</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nb_syst_genes</span><span class="p">)</span>
        <span class="c1"># Nb syst genes matched</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">compute_nb_syst_genes_tot</span><span class="p">())</span>
        <span class="c1"># The total length of the locus in protein number, delimited by hits for profiles of the system.</span>
        <span class="c1">#report_str += &quot;\t{0:d}&quot;.format(self.compute_system_length()) </span>
        <span class="c1"># The total length of the locus in protein number, delimited by hits for profiles of the system.</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">compute_system_length</span><span class="p">(</span><span class="n">rep_info</span><span class="p">))</span>
        <span class="c1"># Nb mandatory_genes matched at least once</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">count_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">))</span>
        <span class="c1"># Nb accessory_genes matched at least once</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">count_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">))</span>

        <span class="n">missing_mandatory</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compute_missing_genes_list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">)</span>
        <span class="n">missing_accessory</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compute_missing_genes_list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">)</span>

        <span class="c1"># Nb mandatory_genes with no occurrence in the system</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">missing_mandatory</span><span class="p">))</span>
        <span class="c1"># Nb accessory_genes with no occurrence in the system</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">missing_accessory</span><span class="p">))</span>
        <span class="c1"># List of mandatory genes with no occurrence in the system</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">missing_mandatory</span><span class="p">)</span>
        <span class="c1"># List of accessory genes with no occurrence in the system</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">missing_accessory</span><span class="p">)</span>
        
        <span class="c1"># The positions of the loci (begin, end) as delimited by hits for profiles of the system.</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">loci_positions</span><span class="p">)</span>
        <span class="c1"># A dico per type of gene &#39;Mandatory, Accessory, Forbidden&#39; with gene occurrences in the system</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_gene_counter_output</span><span class="p">())</span>

        <span class="k">return</span> <span class="n">report_str</span></div>


<div class="viewcode-block" id="SystemOccurence.get_summary_unordered"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.get_summary_unordered">[docs]</a>    <span class="k">def</span> <span class="nf">get_summary_unordered</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">replicon_name</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Gives a summary of the system occurrence in terms of gene content only (specific of &quot;unordered&quot; datasets).</span>

<span class="sd">        :param replicon_name: the name of the replicon</span>
<span class="sd">        :type replicon_name: string</span>
<span class="sd">        :return: a tabulated summary of the :class:`macsypy.search_systems.SystemOccurence`</span>
<span class="sd">        :rtype: string</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1">#report_str = replicon_name+&quot;\t&quot;+self.get_system_unique_name(replicon_name)</span>
        <span class="c1"># No replicon name for unordered... get it from the config object in future developments... </span>
        <span class="n">report_str</span> <span class="o">=</span> <span class="n">replicon_name</span><span class="o">+</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">+</span><span class="bp">self</span><span class="o">.</span><span class="n">get_system_name_unordered</span><span class="p">()</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">)</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">state</span><span class="p">)</span>

        <span class="c1"># Nb of loci included to fill the system occurrence</span>
        <span class="c1">#report_str+=&quot;\t{0:d}&quot;.format(self.nb_cluster) </span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">None&quot;</span><span class="c1"># No loci in unordered</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">))</span> <span class="c1"># Nb mandatory_genes in the definition of the system</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">))</span> <span class="c1"># Nb accessory_genes in the definition of the system</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nb_syst_genes</span><span class="p">)</span> <span class="c1"># Nb syst genes NR</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">compute_nb_syst_genes_tot</span><span class="p">())</span> <span class="c1"># Nb syst genes matched</span>

        <span class="c1"># The total length of the locus in protein number, delimited by hits for profiles of the system.</span>
        <span class="c1">#report_str+=&quot;\t{0:d}&quot;.format(self.compute_system_length(rep_info)) </span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">None&quot;</span> <span class="c1"># No loci in unordered</span>

        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">count_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">))</span> <span class="c1"># Nb mandatory_genes matched at least once</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">count_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">))</span> <span class="c1"># Nb accessory_genes matched at least once</span>

        <span class="n">missing_mandatory</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compute_missing_genes_list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">)</span>        
        <span class="n">missing_accessory</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compute_missing_genes_list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">)</span>

        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">missing_mandatory</span><span class="p">))</span> <span class="c1"># Nb mandatory_genes with no occurrence in the system</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0:d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">missing_accessory</span><span class="p">))</span> <span class="c1"># Nb accessory_genes with no occurrence in the system</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">missing_mandatory</span><span class="p">))</span> <span class="c1"># List of mandatory genes with no occurrence in the system</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">missing_accessory</span><span class="p">))</span> <span class="c1"># List of accessory genes with no occurrence in the system</span>
        
        <span class="c1"># The positions of the loci (begin, end) as delimited by hits for profiles of the system.</span>
        <span class="c1">#report_str+=&quot;\t{0}&quot;.format(self.loci_positions) </span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">None&quot;</span> <span class="c1"># No loci in unordered</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_gene_counter_output</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span> <span class="c1"># A dico per type of gene &#39;Mandatory, Accessory, Forbidden&#39; with gene occurrences in the system</span>

        <span class="k">return</span> <span class="n">report_str</span></div>


<div class="viewcode-block" id="SystemOccurence.fill_with_cluster"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.fill_with_cluster">[docs]</a>    <span class="k">def</span> <span class="nf">fill_with_cluster</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cluster</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Adds hits from a cluster to a system occurence, and check which are their status according to the system definition.</span>
<span class="sd">        Set the system occurence state to &quot;no_decision&quot; after calling of this function.</span>

<span class="sd">        :param cluster: the set of contiguous genes to treat for :class:`macsypy.search_systems.SystemOccurence` inclusion. </span>
<span class="sd">        :type cluster: :class:`macsypy.search_systems.Cluster`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">included</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_state</span> <span class="o">=</span> <span class="s2">&quot;no_decision&quot;</span>
        <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">cluster</span><span class="o">.</span><span class="n">hits</span><span class="p">:</span>
            <span class="c1"># Need to check first that this cluster is eligible for system inclusion</span>
            <span class="c1"># Stores hits for system extraction (positions, sequences) when complete.</span>

            <span class="k">if</span> <span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">is_mandatory</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">system</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">[</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">valid_hit</span> <span class="o">=</span> <span class="n">validSystemHit</span><span class="p">(</span><span class="n">hit</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">,</span> <span class="s2">&quot;mandatory&quot;</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">valid_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">valid_hit</span><span class="p">)</span>
                <span class="c1"># NEW</span>
                <span class="k">if</span> <span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">multi_system</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">multi_syst_genes</span><span class="p">[</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">is_accessory</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">system</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">[</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">valid_hit</span> <span class="o">=</span> <span class="n">validSystemHit</span><span class="p">(</span><span class="n">hit</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">,</span> <span class="s2">&quot;accessory&quot;</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">valid_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">valid_hit</span><span class="p">)</span>
                <span class="c1"># NEW</span>
                <span class="k">if</span> <span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">multi_system</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">multi_syst_genes</span><span class="p">[</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">is_forbidden</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">system</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">forbidden_genes</span><span class="p">[</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">included</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">exmandatory_genes</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">exmandatory_genes</span><span class="p">[</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">]]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">valid_hit</span> <span class="o">=</span> <span class="n">validSystemHit</span><span class="p">(</span><span class="n">hit</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">,</span> <span class="s2">&quot;mandatory&quot;</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">valid_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">valid_hit</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">exaccessory_genes</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">exaccessory_genes</span><span class="p">[</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">]]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">valid_hit</span> <span class="o">=</span> <span class="n">validSystemHit</span><span class="p">(</span><span class="n">hit</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">,</span> <span class="s2">&quot;accessory&quot;</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">valid_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">valid_hit</span><span class="p">)</span>
                <span class="c1"># NEW: exforbidden_genes considered</span>
                <span class="k">elif</span> <span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">exforbidden_genes</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">forbidden_genes</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">exforbidden_genes</span><span class="p">[</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">]]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">valid_hit</span> <span class="o">=</span> <span class="n">validSystemHit</span><span class="p">(</span><span class="n">hit</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">,</span> <span class="s2">&quot;forbidden&quot;</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">valid_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">valid_hit</span><span class="p">)</span>                    
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;Foreign gene </span><span class="si">{0}</span><span class="s2"> in cluster </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">)</span>
                    <span class="c1">#print msg</span>
                    <span class="n">_log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">included</span><span class="p">:</span>
            <span class="c1"># Update the number of loci included in the system</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nb_cluster</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="c1"># Update the positions of the system</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">loci_positions</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">cluster</span><span class="o">.</span><span class="n">begin</span><span class="p">,</span> <span class="n">cluster</span><span class="o">.</span><span class="n">end</span><span class="p">))</span></div>

<div class="viewcode-block" id="SystemOccurence.fill_with_hits"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.fill_with_hits">[docs]</a>    <span class="k">def</span> <span class="nf">fill_with_hits</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">hits</span><span class="p">,</span> <span class="n">include_forbidden</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Adds hits to a system occurence, and check what are their status according to the system definition.</span>
<span class="sd">        Set the system occurence state to &quot;no_decision&quot; after calling of this function.</span>

<span class="sd">        .. note::</span>
<span class="sd">            Forbidden genes will only be included if they do belong to the current system</span>
<span class="sd">            (and not to another specified with &quot;system_ref&quot; in the current system&#39;s definition).</span>

<span class="sd">        :param hits: a list of Hits to treat for :class:`macsypy.search_systems.SystemOccurence` inclusion.</span>
<span class="sd">        :type list of: :class:`macsypy.report.Hit`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_state</span> <span class="o">=</span> <span class="s2">&quot;no_decision&quot;</span>
        <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">hits</span><span class="p">:</span>
            <span class="c1"># Need to check first that this cluster is eligible for system inclusion</span>
            <span class="c1"># Stores hits for system extraction (positions, sequences) when complete. </span>

            <span class="k">if</span> <span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">is_mandatory</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">system</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">[</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">valid_hit</span> <span class="o">=</span> <span class="n">validSystemHit</span><span class="p">(</span><span class="n">hit</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">,</span> <span class="s2">&quot;mandatory&quot;</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">valid_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">valid_hit</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">is_accessory</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">system</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">[</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">valid_hit</span> <span class="o">=</span> <span class="n">validSystemHit</span><span class="p">(</span><span class="n">hit</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">,</span> <span class="s2">&quot;accessory&quot;</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">valid_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">valid_hit</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">is_forbidden</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">system</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">forbidden_genes</span><span class="p">[</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="c1"># SO New: now forbidden genes may be included in the reports:</span>
                <span class="k">if</span> <span class="n">include_forbidden</span><span class="p">:</span>
                    <span class="n">valid_hit</span> <span class="o">=</span> <span class="n">validSystemHit</span><span class="p">(</span><span class="n">hit</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">,</span> <span class="s2">&quot;forbidden&quot;</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">valid_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">valid_hit</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">exmandatory_genes</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">exmandatory_genes</span><span class="p">[</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">]]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">valid_hit</span> <span class="o">=</span> <span class="n">validSystemHit</span><span class="p">(</span><span class="n">hit</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">,</span> <span class="s2">&quot;mandatory&quot;</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">valid_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">valid_hit</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">exaccessory_genes</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">exaccessory_genes</span><span class="p">[</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">]]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">valid_hit</span> <span class="o">=</span> <span class="n">validSystemHit</span><span class="p">(</span><span class="n">hit</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">,</span> <span class="s2">&quot;accessory&quot;</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">valid_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">valid_hit</span><span class="p">)</span>
                <span class="c1"># NEW: exforbidden_genes considered</span>
                <span class="k">elif</span> <span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">exforbidden_genes</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">forbidden_genes</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">exforbidden_genes</span><span class="p">[</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">]]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">valid_hit</span> <span class="o">=</span> <span class="n">validSystemHit</span><span class="p">(</span><span class="n">hit</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">,</span> <span class="s2">&quot;forbidden&quot;</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">valid_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">valid_hit</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;Foreign gene </span><span class="si">{0}</span><span class="s2"> in cluster </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">)</span>
                    <span class="n">_log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span></div>

<div class="viewcode-block" id="SystemOccurence.fill_with_multi_systems_genes"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.fill_with_multi_systems_genes">[docs]</a>    <span class="k">def</span> <span class="nf">fill_with_multi_systems_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">multi_systems_hits</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This function fills the SystemOccurrence with genes putatively coming from other systems (feature &quot;multi_system&quot;).</span>
<span class="sd">        Those genes are used only if the occurrence of the corresponding gene was not yet filled with a gene from a cluster of the system. </span>

<span class="sd">        :param multi_systems_hits: a list of hits of genes that are &quot;multi_system&quot; which correspond to mandatory or accessory genes from the current system for which to fill a SystemOccurrence </span>
<span class="sd">        :type list of: :class:`macsypy.report.Hit`</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># For each &quot;multi_system&quot; gene missing:</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">multi_syst_genes</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">multi_syst_genes</span><span class="p">[</span><span class="n">g</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1">#multi_systems_hits should be a dico gene.name-wise?</span>
                <span class="c1"># We check wether this missing &quot;multi_system&quot; gene was found elsewhere:</span>
                <span class="c1">#if g in multi_gene_names in [multi_gene.name for multi_gene in [hit.gene for hit in multi_systems_hits]]:</span>
                <span class="k">if</span> <span class="n">g</span> <span class="ow">in</span> <span class="p">[</span><span class="n">multi_gene</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">multi_gene</span> <span class="ow">in</span> <span class="p">[</span><span class="n">hit</span><span class="o">.</span><span class="n">gene</span> <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">multi_systems_hits</span><span class="p">]]:</span>
                    <span class="c1"># If so, then the SystemOccurrence is filled with this:</span>
                    <span class="k">if</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">[</span><span class="n">g</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="c1"># Add a valid_hit with a special status? e.g &quot;accessory_multi_system&quot;?</span>
                        <span class="c1">#self.accessory_genes[hit.gene.name]+=1</span>
                        <span class="c1">#valid_hit=validSystemHit(hit, self.system_name, &quot;accessory&quot;)</span>
                        <span class="c1">#self.valid_hits.append(valid_hit)</span>

                    <span class="k">elif</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">[</span><span class="n">g</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="c1"># Add a valid_hit with a special status? e.g &quot;mandatory_multi_system&quot;?</span>
                        <span class="c1">#self.mandatory_genes[hit.gene.name]+=1</span>
                        <span class="c1">#valid_hit=validSystemHit(hit, self.system_name, &quot;mandatory&quot;)</span>
                        <span class="c1">#self.valid_hits.append(valid_hit)</span>

                    <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Gene </span><span class="si">{0}</span><span class="s2"> supplied from a multi_system gene&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="p">))</span></div>
        <span class="c1">#all_hits = [hit for subl in [report.hits for report in all_reports ] for hit in subl]</span>


<div class="viewcode-block" id="SystemOccurence.decision_rule"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.SystemOccurence.decision_rule">[docs]</a>    <span class="k">def</span> <span class="nf">decision_rule</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This function applies the decision rules for system assessment in terms of quorum:</span>
<span class="sd">            - the absence of forbidden genes is checked</span>
<span class="sd">            - the minimal number of mandatory genes is checked (\&quot;min_mandatory_genes_required\&quot;)</span>
<span class="sd">            - the minimal number of genes in the system is checked (\&quot;min_genes_required\&quot;)</span>

<span class="sd">        When a decision is made, the status (self.status) of the </span>
<span class="sd">        :class:`macsypy.search_systems.SystemOccurence` is set either to:</span>

<span class="sd">            - &quot;\single_locus\&quot; when a complete system in the form of a single cluster was found</span>
<span class="sd">            - &quot;\multi_loci\&quot; when a complete system in the form of several clusters was found</span>
<span class="sd">            - &quot;\uncomplete\&quot; when no system was assessed (quorum not reached)</span>
<span class="sd">            - &quot;\empty\&quot; when no gene for this system was found</span>
<span class="sd">            - &quot;\exclude\&quot; when no system was assessed (at least one forbidden gene was found)</span>

<span class="sd">        :return: a printable message of the output decision with this SystemOccurrence</span>
<span class="sd">        :rtype: string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">nb_forbid</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">forbidden_genes</span><span class="p">)</span>
        <span class="n">nb_mandat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mandatory_genes</span><span class="p">)</span>
        <span class="n">nb_accessory</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">accessory_genes</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_nb_syst_genes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compute_nb_syst_genes</span><span class="p">()</span>

        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;====&gt; Decision rule for putative system </span><span class="si">{0}</span><span class="s2">:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">system_name</span><span class="p">)</span>
        <span class="n">msg</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">nb_forbid : </span><span class="si">{0:d}</span><span class="s2"></span>
<span class="s2">nb_mandat : </span><span class="si">{1:d}</span><span class="s2"></span>
<span class="s2">nb_accessory : </span><span class="si">{2:d}</span><span class="s2">&quot;&quot;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_forbid</span><span class="p">,</span> <span class="n">nb_mandat</span><span class="p">,</span> <span class="n">nb_accessory</span><span class="p">)</span>

        <span class="k">if</span> <span class="p">(</span><span class="n">nb_forbid</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">nb_mandat</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">min_mandatory_genes_required</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nb_syst_genes</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">min_genes_required</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nb_syst_genes</span>  <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">):</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">nb_cluster</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> 
                    <span class="bp">self</span><span class="o">.</span><span class="n">_state</span> <span class="o">=</span> <span class="s2">&quot;single_locus&quot;</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_state</span> <span class="o">=</span> <span class="s2">&quot;multi_loci&quot;</span>

                <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Complete </span><span class="se">\&quot;</span><span class="si">{0}</span><span class="se">\&quot;</span><span class="s2"> system.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">state</span><span class="p">)</span>
                <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">******************************************</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="c1">#print msg</span>
                <span class="c1">#_log.info(msg)</span>

            <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">nb_syst_genes</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Uncomplete system.&quot;</span>
                <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">******************************************</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="c1">#print msg</span>
                <span class="c1">#_log.info(msg)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_state</span> <span class="o">=</span> <span class="s2">&quot;uncomplete&quot;</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Empty system.&quot;</span>
                <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">******************************************</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="c1">#print msg</span>
                <span class="c1">#_log.info(msg)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_state</span> <span class="o">=</span> <span class="s2">&quot;empty&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Exclude.&quot;</span>
            <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">******************************************</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="c1">#print msg</span>
            <span class="c1">#_log.info(msg)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_state</span> <span class="o">=</span> <span class="s2">&quot;exclude&quot;</span>

        <span class="k">return</span> <span class="n">msg</span></div></div>

<div class="viewcode-block" id="validSystemHit"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.validSystemHit">[docs]</a><span class="k">class</span> <span class="nc">validSystemHit</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Encapsulates a :class:`macsypy.report.Hit`</span>
<span class="sd">    This class stores a Hit that has been attributed to a detected system. Thus, it also stores:</span>

<span class="sd">    - the system,</span>
<span class="sd">    - the status of the gene in this system,</span>

<span class="sd">    It also aims at storing information for results extraction:</span>

<span class="sd">    - system extraction (e.g. genomic positions)</span>
<span class="sd">    - sequence extraction</span>
<span class="sd">    &quot;&quot;&quot;</span>
<div class="viewcode-block" id="validSystemHit.__init__"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.validSystemHit.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">hit</span><span class="p">,</span> <span class="n">detected_system</span><span class="p">,</span> <span class="n">gene_status</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param hit: a hit to base the validSystemHit on</span>
<span class="sd">        :type hit: :class:`macsypy.report.Hit`</span>
<span class="sd">        :param detected_system: the name of the predicted System</span>
<span class="sd">        :type detected_system: string</span>
<span class="sd">        :param gene_status: the &quot;role&quot; of the gene in the predicted system</span>
<span class="sd">        :type gene_status: string</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_hit</span> <span class="o">=</span> <span class="n">hit</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">predicted_system</span> <span class="o">=</span> <span class="n">detected_system</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference_system</span> <span class="o">=</span> <span class="n">hit</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_status</span> <span class="o">=</span> <span class="n">gene_status</span></div>

    <span class="k">def</span> <span class="nf">__getattr__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">attr_name</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_hit</span><span class="p">,</span> <span class="n">attr_name</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="s2">&quot;</span><span class="si">{id}</span><span class="se">\t</span><span class="si">{rpl_name}</span><span class="se">\t</span><span class="si">{pos:d}</span><span class="se">\t</span><span class="si">{seq_l:d}</span><span class="se">\t</span><span class="si">{gene_name}</span><span class="se">\t</span><span class="si">{ref_sys}</span><span class="se">\t</span><span class="si">{predict_sys}</span><span class="se">\</span>
<span class="se">\t</span><span class="si">{g_status}</span><span class="se">\t</span><span class="si">{i_eval}</span><span class="se">\t</span><span class="si">{score}</span><span class="se">\t</span><span class="si">{prof_cov:f}</span><span class="se">\t</span><span class="si">{seq_cov:f}</span><span class="se">\</span>
<span class="se">\t</span><span class="si">{begin_match:d}</span><span class="se">\t</span><span class="si">{end_match:d}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">id</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                                           <span class="n">rpl_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">replicon_name</span><span class="p">,</span>
                                           <span class="n">pos</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">,</span>
                                           <span class="n">seq_l</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">seq_length</span><span class="p">,</span>
                                           <span class="n">gene_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                                           <span class="n">ref_sys</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_system</span><span class="p">,</span>
                                           <span class="n">predict_sys</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">predicted_system</span><span class="p">,</span>
                                           <span class="n">g_status</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_status</span><span class="p">,</span>
                                           <span class="n">i_eval</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">i_eval</span><span class="p">,</span>
                                           <span class="n">score</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">score</span><span class="p">,</span>
                                           <span class="n">prof_cov</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">profile_coverage</span><span class="p">,</span>
                                           <span class="n">seq_cov</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_coverage</span><span class="p">,</span>
                                           <span class="n">begin_match</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">begin_match</span><span class="p">,</span>
                                           <span class="n">end_match</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">end_match</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">output_system</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">system_name</span><span class="p">,</span> <span class="n">system_status</span><span class="p">):</span>
        <span class="k">return</span> <span class="s2">&quot;</span><span class="si">{id}</span><span class="se">\t</span><span class="si">{rpl_name}</span><span class="se">\t</span><span class="si">{pos:d}</span><span class="se">\t</span><span class="si">{seq_l:d}</span><span class="se">\t</span><span class="si">{gene_name}</span><span class="se">\t</span><span class="si">{ref_sys}</span><span class="se">\t</span><span class="si">{predict_sys}</span><span class="se">\</span>
<span class="s2">        </span><span class="se">\t</span><span class="si">{sys_name}</span><span class="se">\t</span><span class="si">{sys_status}</span><span class="se">\t</span><span class="si">{gene_status}</span><span class="se">\t</span><span class="si">{i_eval:.3e}</span><span class="se">\t</span><span class="si">{score:.3f}</span><span class="se">\t</span><span class="si">{prof_cov:.3f}</span><span class="se">\t</span><span class="si">{seq_cov:.3f}</span><span class="se">\</span>
<span class="s2">        </span><span class="se">\t</span><span class="si">{begin_match:d}</span><span class="se">\t</span><span class="si">{end_match:d}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">id</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                                                   <span class="n">rpl_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">replicon_name</span><span class="p">,</span>
                                                   <span class="n">pos</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">,</span>
                                                   <span class="n">seq_l</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">seq_length</span><span class="p">,</span>
                                                   <span class="n">gene_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                                                   <span class="n">ref_sys</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_system</span><span class="p">,</span>
                                                   <span class="n">predict_sys</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">predicted_system</span><span class="p">,</span>
                                                   <span class="n">sys_name</span><span class="o">=</span><span class="n">system_name</span><span class="p">,</span>
                                                   <span class="n">sys_status</span><span class="o">=</span><span class="n">system_status</span><span class="p">,</span>
                                                   <span class="n">gene_status</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_status</span><span class="p">,</span>
                                                   <span class="n">i_eval</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">i_eval</span><span class="p">,</span>
                                                   <span class="n">score</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">score</span><span class="p">,</span>
                                                   <span class="n">prof_cov</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">profile_coverage</span><span class="p">,</span>
                                                   <span class="n">seq_cov</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_coverage</span><span class="p">,</span>
                                                   <span class="n">begin_match</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">begin_match</span><span class="p">,</span>
                                                   <span class="n">end_match</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">end_match</span><span class="p">)</span>


<div class="viewcode-block" id="validSystemHit.output_system_header"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.validSystemHit.output_system_header">[docs]</a>    <span class="k">def</span> <span class="nf">output_system_header</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :return: the header for the output file</span>
<span class="sd">        :rtype: string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s2">&quot;#Hit_Id</span><span class="se">\t</span><span class="s2">Replicon_name</span><span class="se">\t</span><span class="s2">Position</span><span class="se">\t</span><span class="s2">Sequence_length</span><span class="se">\t</span><span class="s2">Gene</span><span class="se">\t</span><span class="s2">Reference_system</span><span class="se">\t</span><span class="s2">Predicted_system</span><span class="se">\</span>
<span class="se">\t</span><span class="s2">System_Id</span><span class="se">\t</span><span class="s2">System_status</span><span class="se">\t</span><span class="s2">Gene_status</span><span class="se">\t</span><span class="s2">i-evalue</span><span class="se">\t</span><span class="s2">Score</span><span class="se">\t</span><span class="s2">Profile_coverage</span><span class="se">\t</span><span class="s2">Sequence_coverage</span><span class="se">\t</span><span class="s2">Begin_match</span><span class="se">\t</span><span class="s2">End_match</span><span class="se">\n</span><span class="s2">&quot;</span></div></div>



<span class="k">class</span> <span class="nc">systemDetectionReport</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>

    <span class="n">__metaclass__</span> <span class="o">=</span> <span class="n">abc</span><span class="o">.</span><span class="n">ABCMeta</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">systems_occurrences_list</span><span class="p">,</span> <span class="n">cfg</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_systems_occurrences_list</span> <span class="o">=</span> <span class="n">systems_occurrences_list</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cfg</span> <span class="o">=</span> <span class="n">cfg</span>
        <span class="k">if</span> <span class="s1">&#39;MACSY_DEBUG&#39;</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">&#39;MACSY_DEBUG&#39;</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_indent</span> <span class="o">=</span> <span class="mi">2</span> <span class="c1">#human readable json for debugging purpose</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_indent</span> <span class="o">=</span> <span class="kc">None</span> <span class="c1">#improve performance of txssview</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">json_file_name</span> <span class="o">=</span> <span class="s1">&#39;results.macsyfinder.json&#39;</span>


    <span class="nd">@abc</span><span class="o">.</span><span class="n">abstractmethod</span>
    <span class="k">def</span> <span class="nf">report_output</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reportfilename</span><span class="p">,</span> <span class="n">print_header</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Writes a report of sequences forming the detected systems, with information in their status in the system, </span>
<span class="sd">        their localization on replicons, and statistics on the Hits.</span>

<span class="sd">        :param reportfilename: the output file name </span>
<span class="sd">        :type reportfilename: string</span>
<span class="sd">        :param print_header: True if the header has to be written. False otherwise</span>
<span class="sd">        :type print_header: boolean</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">pass</span>

    <span class="nd">@abc</span><span class="o">.</span><span class="n">abstractmethod</span>
    <span class="k">def</span> <span class="nf">summary_output</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reportfilename</span><span class="p">,</span> <span class="n">rep_info</span><span class="p">,</span> <span class="n">print_header</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Writes a report with the summary of systems detected in replicons. For each system, a summary is done including: </span>

<span class="sd">            - the number of mandatory/accessory genes in the reference system (as defined in XML files)</span>
<span class="sd">            - the number of mandatory/accessory genes detected</span>
<span class="sd">            - the number and list of missing genes</span>
<span class="sd">            - the number of loci encoding the system</span>

<span class="sd">        :param rep_info: an entry extracted from the :class:`macsypy.database.RepliconDB`</span>
<span class="sd">        :type rep_info: a namedTuple &quot;RepliconInfo&quot; :class:`macsypy.database.RepliconInfo`</span>
<span class="sd">        :param print_header: True if the header has to be written. False otherwise</span>
<span class="sd">        :type print_header: boolean</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">pass</span>

    <span class="nd">@abc</span><span class="o">.</span><span class="n">abstractmethod</span>
    <span class="k">def</span> <span class="nf">json_output</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">rep_db</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates the report in json format</span>

<span class="sd">        :param path: the path to a file where to write the report in json format</span>
<span class="sd">        :type path: string</span>
<span class="sd">        :param rep_db: the replicon database</span>
<span class="sd">        :type rep_db: a class:`macsypy.database.RepliconDB` object</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">pass</span>


<div class="viewcode-block" id="systemDetectionReportOrdered"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.systemDetectionReportOrdered">[docs]</a><span class="k">class</span> <span class="nc">systemDetectionReportOrdered</span><span class="p">(</span><span class="n">systemDetectionReport</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Stores the detected systems to report for each replicon: </span>
<span class="sd">        - by system name, </span>
<span class="sd">        - by state of the systems (single vs multi loci)</span>

<span class="sd">    &quot;&quot;&quot;</span>


<div class="viewcode-block" id="systemDetectionReportOrdered.__init__"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.systemDetectionReportOrdered.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">replicon_name</span><span class="p">,</span> <span class="n">systems_occurrences_list</span><span class="p">,</span> <span class="n">cfg</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param replicon_name: the name of the replicon</span>
<span class="sd">        :type replicon_name: string</span>
<span class="sd">        :param systems_occurrences_list: the list of system&#39;s occurrences to consider</span>
<span class="sd">        :type systems_occurrences_list: list of :class:`macsypy.search_systems.SystemOccurence`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">systemDetectionReportOrdered</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">systems_occurrences_list</span><span class="p">,</span> <span class="n">cfg</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">replicon_name</span> <span class="o">=</span> <span class="n">replicon_name</span></div>



<div class="viewcode-block" id="systemDetectionReportOrdered.counter_output"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.systemDetectionReportOrdered.counter_output">[docs]</a>    <span class="k">def</span> <span class="nf">counter_output</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Builds a counter of systems per replicon, with different &quot;states&quot; separated (single-locus vs multi-loci systems)</span>

<span class="sd">        :return: the counter of systems</span>
<span class="sd">        :rtype: Counter</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">system_textlist</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">so</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_systems_occurrences_list</span><span class="p">:</span>
            <span class="n">system_textlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system_name</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">so</span><span class="o">.</span><span class="n">state</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">Counter</span><span class="p">(</span><span class="n">system_textlist</span><span class="p">)</span></div>

<div class="viewcode-block" id="systemDetectionReportOrdered.tabulated_output_header"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.systemDetectionReportOrdered.tabulated_output_header">[docs]</a>    <span class="k">def</span> <span class="nf">tabulated_output_header</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">system_occurrence_states</span><span class="p">,</span> <span class="n">system_names</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a string containing the header of the tabulated output</span>

<span class="sd">        :param system_occurrence_states: the different forms of detected systems to consider</span>
<span class="sd">        :type system_occurrence_states: list of string</span>
<span class="sd">        :rtype: string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Can be done intra-class </span>
        <span class="n">header</span> <span class="o">=</span> <span class="s2">&quot;#Replicon&quot;</span>
        <span class="k">for</span> <span class="n">syst_name</span> <span class="ow">in</span> <span class="n">system_names</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">state</span> <span class="ow">in</span> <span class="n">system_occurrence_states</span><span class="p">:</span>
                <span class="n">header</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">syst_name</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">state</span>
        <span class="n">header</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">return</span> <span class="n">header</span></div>

<div class="viewcode-block" id="systemDetectionReportOrdered.tabulated_output"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.systemDetectionReportOrdered.tabulated_output">[docs]</a>    <span class="k">def</span> <span class="nf">tabulated_output</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">system_occurrence_states</span><span class="p">,</span> <span class="n">system_names</span><span class="p">,</span> <span class="n">reportfilename</span><span class="p">,</span> <span class="n">print_header</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Write a tabulated output with number of detected systems for each replicon.</span>

<span class="sd">        :param system_occurrence_states: the different forms of detected systems to consider</span>
<span class="sd">        :type system_occurrence_states: list of string</span>
<span class="sd">        :param reportfilename: the output file name </span>
<span class="sd">        :type reportfilename: string</span>
<span class="sd">        :param print_header: True if the header has to be written. False otherwise</span>
<span class="sd">        :type print_header: boolean</span>
<span class="sd">        :rtype: string</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">system_counter</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">counter_output</span><span class="p">()</span>
        <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">system_counter</span><span class="p">)</span>
        <span class="n">report_str</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">replicon_name</span>
        <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">system_names</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">o</span> <span class="ow">in</span> <span class="n">system_occurrence_states</span><span class="p">:</span>
                <span class="n">index</span> <span class="o">=</span> <span class="n">s</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">o</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">index</span> <span class="ow">in</span> <span class="n">system_counter</span><span class="p">:</span>
                    <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span>
                    <span class="n">report_str</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="n">system_counter</span><span class="p">[</span><span class="n">index</span><span class="p">])</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">0&quot;</span>
        <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">reportfilename</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">_file</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">print_header</span><span class="p">:</span>
                <span class="n">_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tabulated_output_header</span><span class="p">(</span><span class="n">system_occurrence_states</span><span class="p">,</span> <span class="n">system_names</span><span class="p">))</span>
            <span class="n">_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">report_str</span><span class="p">)</span></div>


<div class="viewcode-block" id="systemDetectionReportOrdered.report_output"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.systemDetectionReportOrdered.report_output">[docs]</a>    <span class="k">def</span> <span class="nf">report_output</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reportfilename</span><span class="p">,</span> <span class="n">print_header</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Writes a report of sequences forming the detected systems, with information in their status in the system, </span>
<span class="sd">        their localization on replicons, and statistics on the Hits.</span>

<span class="sd">        :param reportfilename: the output file name </span>
<span class="sd">        :type reportfilename: string</span>
<span class="sd">        :param print_header: True if the header has to be written. False otherwise</span>
<span class="sd">        :type print_header: boolean</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">report_str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">for</span> <span class="n">so</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_systems_occurrences_list</span><span class="p">:</span>
            <span class="n">so_unique_name</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">get_system_unique_name</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">replicon_name</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">so</span><span class="o">.</span><span class="n">valid_hits</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">print_header</span><span class="p">:</span>
                    <span class="n">report_str</span> <span class="o">+=</span> <span class="n">hit</span><span class="o">.</span><span class="n">output_system_header</span><span class="p">()</span>
                    <span class="n">print_header</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="n">report_str</span> <span class="o">+=</span> <span class="n">hit</span><span class="o">.</span><span class="n">output_system</span><span class="p">(</span><span class="n">so_unique_name</span><span class="p">,</span> <span class="n">so</span><span class="o">.</span><span class="n">state</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">reportfilename</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">_file</span><span class="p">:</span>
            <span class="n">_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">report_str</span><span class="p">)</span></div>


<div class="viewcode-block" id="systemDetectionReportOrdered.summary_output"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.systemDetectionReportOrdered.summary_output">[docs]</a>    <span class="k">def</span> <span class="nf">summary_output</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reportfilename</span><span class="p">,</span> <span class="n">rep_info</span><span class="p">,</span> <span class="n">print_header</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Writes a report with the summary of systems detected in replicons. For each system, a summary is done including:</span>

<span class="sd">            - the number of mandatory/accessory genes in the reference system (as defined in XML files)</span>
<span class="sd">            - the number of mandatory/accessory genes detected</span>
<span class="sd">            - the number and list of missing genes</span>
<span class="sd">            - the number of loci encoding the system</span>

<span class="sd">        :param rep_info: an entry extracted from the :class:`macsypy.database.RepliconDB`</span>
<span class="sd">        :type rep_info: a namedTuple &quot;RepliconInfo&quot; :class:`macsypy.database.RepliconInfo`</span>
<span class="sd">        :param print_header: True if the header has to be written. False otherwise</span>
<span class="sd">        :type print_header: boolean</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">report_str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">for</span> <span class="n">so</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_systems_occurrences_list</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">print_header</span><span class="p">:</span>
                <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">get_summary_header</span><span class="p">())</span>
                <span class="n">print_header</span> <span class="o">=</span> <span class="kc">False</span>

            <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">get_summary</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">replicon_name</span><span class="p">,</span> <span class="n">rep_info</span><span class="p">))</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">reportfilename</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">_file</span><span class="p">:</span>
            <span class="n">_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">report_str</span><span class="p">)</span></div>


<div class="viewcode-block" id="systemDetectionReportOrdered.json_output"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.systemDetectionReportOrdered.json_output">[docs]</a>    <span class="k">def</span> <span class="nf">json_output</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">json_path</span><span class="p">,</span> <span class="n">json_data</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">json_path</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">_file</span><span class="p">:</span>
            <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">json_data</span><span class="p">,</span> <span class="n">_file</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_indent</span><span class="p">)</span></div>


<div class="viewcode-block" id="systemDetectionReportOrdered._match2json"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.systemDetectionReportOrdered._match2json">[docs]</a>    <span class="k">def</span> <span class="nf">_match2json</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">valid_hit</span><span class="p">,</span> <span class="n">so</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param valid_hit: the valid hit to transform in to json.</span>
<span class="sd">        :type valid_hit: class:`macsypy.search_system.ValidHit` object.</span>
<span class="sd">        :param so: the system occurence where the valid hit come from.</span>
<span class="sd">        :type so: class:`macsypy.search_system.SystemOccurence.`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">id</span>
        <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">position</span>
        <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;sequence_length&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">seq_length</span>
        <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;system&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">reference_system</span>
        <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;match&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span>
        <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;gene_status&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">gene_status</span>
        <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;i_eval&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">i_eval</span>
        <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">score</span>
        <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;profile_coverage&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">profile_coverage</span>
        <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;sequence_coverage&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">sequence_coverage</span>
        <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;begin_match&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">begin_match</span>
        <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;end_match&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">end_match</span>
        <span class="n">gene_ref</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">get_gene_ref</span><span class="p">(</span><span class="n">valid_hit</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">gene_ref</span><span class="p">:</span>
            <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;function&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_ref</span><span class="o">.</span><span class="n">name</span>
        <span class="k">return</span> <span class="n">gene</span></div>

    <span class="k">def</span> <span class="nf">_gene2json</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">sequence_length</span><span class="p">,</span> <span class="n">position</span><span class="p">):</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="n">gene_name</span><span class="p">,</span>
                <span class="s1">&#39;sequence_length&#39;</span> <span class="p">:</span> <span class="n">sequence_length</span><span class="p">,</span>
                <span class="s1">&#39;position&#39;</span><span class="p">:</span> <span class="n">position</span>
                <span class="p">}</span>
        <span class="k">return</span> <span class="n">gene</span>


<div class="viewcode-block" id="systemDetectionReportOrdered.system_2_json"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.systemDetectionReportOrdered.system_2_json">[docs]</a>    <span class="k">def</span> <span class="nf">system_2_json</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rep_db</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates the report in json format</span>

<span class="sd">        :param path: the path to a file where to write the report in json format</span>
<span class="sd">        :type path: string</span>
<span class="sd">        :param rep_db: the replicon database</span>
<span class="sd">        :type rep_db: a class:`macsypy.database.RepliconDB` object</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">systems</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">so</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_systems_occurrences_list</span><span class="p">:</span>
            <span class="n">system</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">system_name</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">system_name</span>
            <span class="n">fields</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">unique_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>
            <span class="n">repliconName</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">occurrence_number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
            
            <span class="n">system</span><span class="p">[</span><span class="s1">&#39;occurrence_number&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">occurrence_number</span>
            <span class="n">system</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">system_name</span>
            <span class="n">system</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">unique_name</span>
            <span class="n">system</span><span class="p">[</span><span class="s1">&#39;replicon&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">system</span><span class="p">[</span><span class="s1">&#39;replicon&#39;</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">valid_hits</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replicon_name</span> <span class="c1"># Ok, Otherwise the object has a field self.replicon_name</span>
            <span class="n">rep_info</span> <span class="o">=</span> <span class="n">rep_db</span><span class="p">[</span><span class="n">system</span><span class="p">[</span><span class="s1">&#39;replicon&#39;</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">]]</span>
            <span class="n">system</span><span class="p">[</span><span class="s1">&#39;replicon&#39;</span><span class="p">][</span><span class="s1">&#39;length&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">max</span> <span class="o">-</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">min</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">system</span><span class="p">[</span><span class="s1">&#39;replicon&#39;</span><span class="p">][</span><span class="s1">&#39;topology&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">topology</span>
            <span class="n">system</span><span class="p">[</span><span class="s1">&#39;genes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">if</span> <span class="n">so</span><span class="o">.</span><span class="n">valid_hits</span><span class="p">:</span>
                <span class="n">positions</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span><span class="o">.</span><span class="n">position</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">so</span><span class="o">.</span><span class="n">valid_hits</span><span class="p">]</span>
                <span class="n">valid_hits</span> <span class="o">=</span> <span class="p">{</span><span class="n">vh</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="n">vh</span> <span class="k">for</span> <span class="n">vh</span> <span class="ow">in</span> <span class="n">so</span><span class="o">.</span><span class="n">valid_hits</span><span class="p">}</span>
                <span class="n">pos_min</span> <span class="o">=</span> <span class="n">positions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">5</span>
                <span class="k">if</span> <span class="n">pos_min</span> <span class="o">&lt;</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">min</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">topology</span> <span class="o">==</span> <span class="s1">&#39;circular&#39;</span><span class="p">:</span>
                        <span class="n">pos_min</span> <span class="o">=</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">max</span> <span class="o">+</span> <span class="n">positions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">5</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">pos_min</span> <span class="o">=</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">min</span>
                <span class="n">pos_max</span> <span class="o">=</span> <span class="n">positions</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">5</span>
                <span class="k">if</span> <span class="n">pos_max</span> <span class="o">&gt;</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">max</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">topology</span> <span class="o">==</span> <span class="s1">&#39;circular&#39;</span><span class="p">:</span>
                        <span class="n">pos_max</span> <span class="o">=</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">max</span> <span class="o">-</span> <span class="n">positions</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">5</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">pos_max</span> <span class="o">=</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">max</span>
                <span class="k">if</span> <span class="n">pos_min</span> <span class="o">&lt;</span> <span class="n">pos_max</span><span class="p">:</span>
                    <span class="n">pos_in_bk_2_display</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="n">pos_min</span><span class="p">,</span> <span class="n">pos_max</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">before_orig</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="n">pos_min</span><span class="p">,</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">max</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
                    <span class="n">after_orig</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="n">rep_info</span><span class="o">.</span><span class="n">min</span><span class="p">,</span> <span class="n">pos_max</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
                    <span class="n">pos_in_bk_2_display</span> <span class="o">=</span> <span class="n">before_orig</span> <span class="o">+</span> <span class="n">after_orig</span>
                <span class="n">pos_in_rep_2_display</span> <span class="o">=</span> <span class="p">[</span><span class="n">pos</span> <span class="o">-</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">min</span> <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">pos_in_bk_2_display</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">curr_position</span> <span class="ow">in</span> <span class="n">pos_in_rep_2_display</span><span class="p">:</span>
                    <span class="n">gene_name</span><span class="p">,</span> <span class="n">gene_length</span> <span class="o">=</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">curr_position</span><span class="p">]</span>
                    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">cfg</span><span class="o">.</span><span class="n">db_type</span> <span class="o">==</span> <span class="s1">&#39;gembase&#39;</span><span class="p">:</span>
                        <span class="c1"># SO - PB WAS HERE, NAMES WERE WRONG after the 1st replicon. Thus the gene_id is NEVER in the valid_hits.</span>
                        <span class="n">gene_id</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">system</span><span class="p">[</span><span class="s1">&#39;replicon&#39;</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">],</span> <span class="n">gene_name</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">gene_id</span> <span class="o">=</span> <span class="n">gene_name</span>
                    <span class="k">if</span> <span class="n">gene_id</span> <span class="ow">in</span> <span class="n">valid_hits</span><span class="p">:</span>
                        <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_match2json</span><span class="p">(</span><span class="n">valid_hits</span><span class="p">[</span><span class="n">gene_id</span><span class="p">],</span> <span class="n">so</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_gene2json</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">gene_length</span><span class="p">),</span> <span class="n">curr_position</span> <span class="o">+</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">min</span><span class="p">)</span>
                    <span class="n">system</span><span class="p">[</span><span class="s1">&#39;genes&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

            <span class="n">system</span><span class="p">[</span><span class="s1">&#39;summary&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">system</span><span class="p">[</span><span class="s1">&#39;summary&#39;</span><span class="p">][</span><span class="s1">&#39;mandatory&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">mandatory_genes</span>
            <span class="n">system</span><span class="p">[</span><span class="s1">&#39;summary&#39;</span><span class="p">][</span><span class="s1">&#39;accessory&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">accessory_genes</span>
            <span class="n">system</span><span class="p">[</span><span class="s1">&#39;summary&#39;</span><span class="p">][</span><span class="s1">&#39;forbidden&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">forbidden_genes</span>
            <span class="n">system</span><span class="p">[</span><span class="s1">&#39;summary&#39;</span><span class="p">][</span><span class="s1">&#39;state&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">_state</span>
            <span class="n">systems</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">system</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">systems</span></div></div>


<div class="viewcode-block" id="systemDetectionReportUnordered"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.systemDetectionReportUnordered">[docs]</a><span class="k">class</span> <span class="nc">systemDetectionReportUnordered</span><span class="p">(</span><span class="n">systemDetectionReport</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Stores a report for putative detected systems gathering all hits from a search in an unordered dataset: </span>
<span class="sd">        - by system.</span>

<span class="sd">    Mandatory and accessory genes only are reported in the &quot;json&quot; and &quot;report&quot; output,</span>
<span class="sd">    but all hits matching a system component are reported in the &quot;summary&quot;.</span>

<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="systemDetectionReportUnordered.__init__"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.systemDetectionReportUnordered.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">systems_occurrences_list</span><span class="p">,</span> <span class="n">cfg</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param systems_occurrences_list: the list of system&#39;s occurrences to consider</span>
<span class="sd">        :type systems_occurrences_list: list of :class:`macsypy.search_systems.SystemOccurence`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">systemDetectionReportUnordered</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">systems_occurrences_list</span><span class="p">,</span> <span class="n">cfg</span><span class="p">)</span></div>



<div class="viewcode-block" id="systemDetectionReportUnordered.report_output"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.systemDetectionReportUnordered.report_output">[docs]</a>    <span class="k">def</span> <span class="nf">report_output</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reportfilename</span><span class="p">,</span> <span class="n">print_header</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Writes a report of sequences forming the detected systems, with information in their status in the system, </span>
<span class="sd">        their localization on replicons, and statistics on the Hits.</span>

<span class="sd">        :param reportfilename: the output file name </span>
<span class="sd">        :type reportfilename: string</span>
<span class="sd">        :param print_header: True if the header has to be written. False otherwise</span>
<span class="sd">        :type print_header: boolean</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">report_str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">for</span> <span class="n">so</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_systems_occurrences_list</span><span class="p">:</span>
            <span class="c1">#so_unique_name = so.get_system_unique_name(self.replicon_name)</span>
            <span class="n">so_unique_name</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">get_system_name_unordered</span><span class="p">()</span>
            <span class="c1">#so_unique_name = so.system_name+&quot;_putative&quot;</span>
            <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">so</span><span class="o">.</span><span class="n">valid_hits</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">print_header</span><span class="p">:</span>
                    <span class="n">report_str</span> <span class="o">+=</span> <span class="n">hit</span><span class="o">.</span><span class="n">output_system_header</span><span class="p">()</span>
                    <span class="n">print_header</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="n">report_str</span> <span class="o">+=</span> <span class="n">hit</span><span class="o">.</span><span class="n">output_system</span><span class="p">(</span><span class="n">so_unique_name</span><span class="p">,</span> <span class="n">so</span><span class="o">.</span><span class="n">state</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">reportfilename</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">_file</span><span class="p">:</span>
            <span class="n">_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">report_str</span><span class="p">)</span></div>


<div class="viewcode-block" id="systemDetectionReportUnordered.summary_output"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.systemDetectionReportUnordered.summary_output">[docs]</a>    <span class="k">def</span> <span class="nf">summary_output</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reportfilename</span><span class="p">,</span> <span class="n">print_header</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Writes a report with the summary for putative systems in an unordered dataset. For each system, a summary is done including:</span>

<span class="sd">            - the number of mandatory/accessory genes in the reference system (as defined in XML files)</span>
<span class="sd">            - the number of mandatory/accessory genes detected</span>

<span class="sd">        :param reportfilename: the output file name </span>
<span class="sd">        :type reportfilename: string</span>
<span class="sd">        :param print_header: True if the header has to be written. False otherwise</span>
<span class="sd">        :type print_header: boolean</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">report_str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">for</span> <span class="n">so</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_systems_occurrences_list</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">print_header</span><span class="p">:</span>
                <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">get_summary_header</span><span class="p">())</span>
                <span class="n">print_header</span> <span class="o">=</span> <span class="kc">False</span>

            <span class="c1">#report_str+=&quot;{0}\n&quot;.format(so.get_summary(self.replicon_name, rep_info))</span>
            <span class="c1"># Get a fake &quot;replicon_name&quot; from the config object in future devt.</span>
            <span class="n">report_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">get_summary_unordered</span><span class="p">(</span><span class="s2">&quot;Unordered&quot;</span><span class="p">))</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">reportfilename</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">_file</span><span class="p">:</span>
            <span class="n">_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">report_str</span><span class="p">)</span></div>



<div class="viewcode-block" id="systemDetectionReportUnordered.json_output"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.systemDetectionReportUnordered.json_output">[docs]</a>    <span class="k">def</span> <span class="nf">json_output</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">json_path</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates the report in json format</span>

<span class="sd">        :param path: the path to a file where to write the report in json format</span>
<span class="sd">        :type path: string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">def</span> <span class="nf">cmp_so</span><span class="p">(</span><span class="n">so</span><span class="p">,</span> <span class="n">vh_1</span><span class="p">,</span> <span class="n">vh_2</span><span class="p">):</span>
            <span class="n">gene_1</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">get_gene_ref</span><span class="p">(</span><span class="n">vh_1</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">gene_1</span><span class="p">:</span>
                <span class="n">gene_1</span> <span class="o">=</span> <span class="n">vh_1</span><span class="o">.</span><span class="n">gene</span>
            <span class="n">gene_2</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">get_gene_ref</span><span class="p">(</span><span class="n">vh_2</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">gene_2</span><span class="p">:</span>
                <span class="n">gene_2</span> <span class="o">=</span> <span class="n">vh_2</span><span class="o">.</span><span class="n">gene</span>
            <span class="k">if</span> <span class="n">gene_1</span><span class="o">.</span><span class="n">is_mandatory</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">)</span> <span class="ow">and</span> <span class="n">gene_2</span><span class="o">.</span><span class="n">is_mandatory</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">):</span>
                <span class="k">return</span> <span class="nb">cmp</span><span class="p">(</span><span class="n">vh_1</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">vh_2</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">gene_1</span><span class="o">.</span><span class="n">is_mandatory</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">)</span> <span class="ow">and</span> <span class="n">gene_2</span><span class="o">.</span><span class="n">is_accessory</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">):</span>
                <span class="k">return</span> <span class="o">-</span><span class="mi">1</span>
            <span class="k">elif</span> <span class="n">gene_1</span><span class="o">.</span><span class="n">is_mandatory</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">)</span> <span class="ow">and</span> <span class="n">gene_2</span><span class="o">.</span><span class="n">is_forbidden</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">):</span>
                <span class="k">return</span> <span class="o">-</span><span class="mi">1</span>

            <span class="k">elif</span> <span class="n">gene_1</span><span class="o">.</span><span class="n">is_accessory</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">)</span> <span class="ow">and</span> <span class="n">gene_2</span><span class="o">.</span><span class="n">is_mandatory</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">):</span>
                <span class="k">return</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="n">gene_1</span><span class="o">.</span><span class="n">is_accessory</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">)</span> <span class="ow">and</span> <span class="n">gene_2</span><span class="o">.</span><span class="n">is_accessory</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">):</span>
                <span class="k">return</span> <span class="nb">cmp</span><span class="p">(</span><span class="n">vh_1</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">vh_2</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">gene_1</span><span class="o">.</span><span class="n">is_accessory</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">)</span> <span class="ow">and</span> <span class="n">gene_2</span><span class="o">.</span><span class="n">is_forbidden</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">):</span>
                <span class="k">return</span> <span class="o">-</span><span class="mi">1</span>

            <span class="k">elif</span> <span class="n">gene_1</span><span class="o">.</span><span class="n">is_forbidden</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">)</span> <span class="ow">and</span> <span class="n">gene_2</span><span class="o">.</span><span class="n">is_mandatory</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">):</span>
                <span class="k">return</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="n">gene_1</span><span class="o">.</span><span class="n">is_forbidden</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">)</span> <span class="ow">and</span> <span class="n">gene_2</span><span class="o">.</span><span class="n">is_accessory</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">):</span>
                <span class="k">return</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="n">gene_1</span><span class="o">.</span><span class="n">is_forbidden</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">)</span> <span class="ow">and</span> <span class="n">gene_2</span><span class="o">.</span><span class="n">is_forbidden</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">system</span><span class="p">):</span>
                <span class="k">return</span> <span class="nb">cmp</span><span class="p">(</span><span class="n">vh_1</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">vh_2</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
            <span class="k">assert</span> <span class="kc">False</span><span class="p">,</span> <span class="s2">&quot;problem during hit comparison&quot;</span>


        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">json_path</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">_file</span><span class="p">:</span>
            <span class="n">systems</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">so</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_systems_occurrences_list</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">so</span><span class="o">.</span><span class="n">unique_name</span><span class="p">:</span>
                    <span class="n">so</span><span class="o">.</span><span class="n">unique_name</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">get_system_name_unordered</span><span class="p">()</span>
                <span class="n">system</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="n">system</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">system_name</span>
                <span class="n">system</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">unique_name</span>
                <span class="n">system</span><span class="p">[</span><span class="s1">&#39;replicon&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="n">system</span><span class="p">[</span><span class="s1">&#39;replicon&#39;</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">valid_hits</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replicon_name</span> <span class="c1"># Ok, Otherwise the object has a field self.replicon_name</span>
                <span class="n">system</span><span class="p">[</span><span class="s1">&#39;genes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">so</span><span class="o">.</span><span class="n">valid_hits</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="nb">cmp</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span><span class="n">cmp_so</span><span class="p">(</span><span class="n">so</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">))</span>
                <span class="k">for</span> <span class="n">valid_hit</span> <span class="ow">in</span> <span class="n">so</span><span class="o">.</span><span class="n">valid_hits</span><span class="p">:</span>
                    <span class="n">gene</span> <span class="o">=</span> <span class="p">{}</span>
                    <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">id</span>
                    <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">position</span>
                    <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;sequence_length&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">seq_length</span>
                    <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;system&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">reference_system</span>
                    <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;match&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span>
                    <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;gene_status&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">gene_status</span>
                    <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;i_eval&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">i_eval</span>
                    <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">score</span>
                    <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;profile_coverage&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">profile_coverage</span>
                    <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;sequence_coverage&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">sequence_coverage</span>
                    <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;begin_match&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">begin_match</span>
                    <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;end_match&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">valid_hit</span><span class="o">.</span><span class="n">end_match</span>
                    <span class="n">gene_ref</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">get_gene_ref</span><span class="p">(</span><span class="n">valid_hit</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">gene_ref</span><span class="p">:</span>
                        <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;function&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_ref</span><span class="o">.</span><span class="n">name</span>
                    <span class="n">system</span><span class="p">[</span><span class="s1">&#39;genes&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                <span class="n">system</span><span class="p">[</span><span class="s1">&#39;summary&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="n">system</span><span class="p">[</span><span class="s1">&#39;summary&#39;</span><span class="p">][</span><span class="s1">&#39;mandatory&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">mandatory_genes</span>
                <span class="n">system</span><span class="p">[</span><span class="s1">&#39;summary&#39;</span><span class="p">][</span><span class="s1">&#39;accessory&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">accessory_genes</span>
                <span class="n">system</span><span class="p">[</span><span class="s1">&#39;summary&#39;</span><span class="p">][</span><span class="s1">&#39;forbidden&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">forbidden_genes</span>
                <span class="n">system</span><span class="p">[</span><span class="s1">&#39;summary&#39;</span><span class="p">][</span><span class="s1">&#39;state&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">_state</span>
                <span class="n">systems</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">system</span><span class="p">)</span>
            <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">systems</span><span class="p">,</span> <span class="n">_file</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_indent</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="get_compatible_systems"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.get_compatible_systems">[docs]</a><span class="k">def</span> <span class="nf">get_compatible_systems</span><span class="p">(</span><span class="n">systems_list1</span><span class="p">,</span> <span class="n">systems_list2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns the intersection of the two input systems lists.</span>

<span class="sd">    :param systems_list1, systems_list2: two lists of systems </span>
<span class="sd">    :type systems_list1, systems_list2: two lists of :class:`macsypy.system.System`</span>
<span class="sd">    :return: a list of systems, or an empty list if no common system</span>
<span class="sd">    :rtype: a list of :class:`macsypy.system.System`</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">inter</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">el</span> <span class="ow">in</span> <span class="n">systems_list1</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">el</span> <span class="ow">in</span> <span class="n">systems_list2</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">el</span> <span class="ow">in</span> <span class="n">inter</span><span class="p">:</span>
                <span class="n">inter</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">el</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">inter</span></div>


<div class="viewcode-block" id="disambiguate_cluster"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.disambiguate_cluster">[docs]</a><span class="k">def</span> <span class="nf">disambiguate_cluster</span><span class="p">(</span><span class="n">cluster</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This disambiguation step is used on clusters with hits for multiple systems</span>
<span class="sd">    (when cluster.state is set to &quot;ambiguous&quot;). It returns a &quot;cleansed&quot; list of clusters,</span>
<span class="sd">    ready to use for system occurence detection (and that are &quot;clear&quot; cases). It:</span>

<span class="sd">    - splits the cluster in two if it seems that two systems are nearby</span>
<span class="sd">    - removes single hits that are not forbidden for the &quot;main&quot; system and</span>
<span class="sd">      that are at one end of the current cluster in this case,</span>
<span class="sd">      check that they are not &quot;loners&quot;, cause &quot;loners&quot; can be stored.</span>

<span class="sd">    :param cluster: the cluster to &quot;disambiguate&quot;</span>
<span class="sd">    :type cluster: :class:`macsypy.search_systems.Cluster`</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">res_clusters</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">counter_genes_compat_systems</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Disambiguation step:&quot;</span><span class="p">)</span>

    <span class="n">cur_cluster</span> <span class="o">=</span> <span class="n">Cluster</span><span class="p">(</span><span class="n">cluster</span><span class="o">.</span><span class="n">systems_to_detect</span><span class="p">)</span> <span class="c1"># New</span>
    <span class="n">cur_cluster</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">cluster</span><span class="o">.</span><span class="n">hits</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1"># Now more complex, deals with compatible systems also for disambiguation.</span>
    <span class="n">cur_compatible</span> <span class="o">=</span> <span class="n">cluster</span><span class="o">.</span><span class="n">hits</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">get_compatible_systems</span><span class="p">(</span><span class="n">cluster</span><span class="o">.</span><span class="n">systems_to_detect</span><span class="p">)</span>

    <span class="c1">#print cluster.hits[0]</span>
    <span class="c1">#print [syst.name for syst in cur_compatible]</span>
    <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">cluster</span><span class="o">.</span><span class="n">hits</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
        <span class="c1">#compatible_systems = h.gene.get_compatible_systems(cluster.systems_to_detect) # tmp before yep</span>
        <span class="n">compatible_systems</span> <span class="o">=</span> <span class="n">h</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">get_compatible_systems</span><span class="p">(</span><span class="n">cluster</span><span class="o">.</span><span class="n">systems_to_detect</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># tmp before nope</span>
        <span class="n">compat_list</span> <span class="o">=</span> <span class="n">get_compatible_systems</span><span class="p">(</span><span class="n">cur_compatible</span><span class="p">,</span> <span class="n">compatible_systems</span><span class="p">)</span> <span class="c1"># intersection for the two genes to agglomerate</span>

        <span class="c1">#print h</span>
        <span class="c1">#print &quot;Hit&#39;s:&quot;</span>
        <span class="c1">#print [syst.name for syst in compatible_systems]</span>
        <span class="c1">#print &quot;Inter:&quot;</span>
        <span class="c1">#print [syst.name for syst in compat_list]</span>
        <span class="k">if</span> <span class="n">compat_list</span><span class="p">:</span>
            <span class="c1"># The two consecutive genes have at least one common compatible system</span>
            <span class="n">cur_cluster</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>
            <span class="n">cur_compatible</span> <span class="o">=</span> <span class="n">compat_list</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Deal with &quot;accessory foreign genes&quot;,</span>
            <span class="c1"># and system attribution when the current gene can be in both aside systems!</span>
            <span class="c1"># Former cluster (now to save) is considered.</span>
            <span class="c1"># Check cluster status before storing it or not:</span>
            <span class="n">cur_cluster</span><span class="o">.</span><span class="n">save</span><span class="p">()</span> <span class="c1"># Check if it updates compatible systems?? right?</span>
            <span class="k">if</span> <span class="n">cur_cluster</span><span class="o">.</span><span class="n">state</span> <span class="o">==</span> <span class="s2">&quot;clear&quot;</span><span class="p">:</span>
                <span class="c1"># Update counts of compatible systems with the current cluster to store</span>
                <span class="c1">#print &quot;\nclear to store: &quot;</span>
                <span class="c1">#print cur_cluster</span>
                <span class="n">res_clusters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_cluster</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">syst</span> <span class="ow">in</span> <span class="n">cur_compatible</span><span class="p">:</span> <span class="c1"># Good list of compatible??</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">counter_genes_compat_systems</span><span class="o">.</span><span class="n">has_key</span><span class="p">(</span><span class="n">syst</span><span class="o">.</span><span class="n">name</span><span class="p">):</span>
                        <span class="n">counter_genes_compat_systems</span><span class="p">[</span><span class="n">syst</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cur_cluster</span><span class="o">.</span><span class="n">hits</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">counter_genes_compat_systems</span><span class="p">[</span><span class="n">syst</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cur_cluster</span><span class="o">.</span><span class="n">hits</span><span class="p">)</span>
                    <span class="c1">#print counter_genes_compat_systems</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Update counts of compatibles systems for all the hits</span>
                <span class="c1">#print &quot;\nnope to store: &quot;</span>
                <span class="c1">#print cur_cluster</span>
                <span class="k">for</span> <span class="n">h_clust</span> <span class="ow">in</span> <span class="n">cur_cluster</span><span class="o">.</span><span class="n">hits</span><span class="p">:</span>
                    <span class="c1">#h_compat = h_clust.gene.get_compatible_systems(cluster.systems_to_detect) # tmp before yep</span>
                    <span class="n">h_compat</span> <span class="o">=</span> <span class="n">h_clust</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">get_compatible_systems</span><span class="p">(</span><span class="n">cluster</span><span class="o">.</span><span class="n">systems_to_detect</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># tmp before nope</span>
                    <span class="k">for</span> <span class="n">syst</span> <span class="ow">in</span> <span class="n">h_compat</span><span class="p">:</span>
                        <span class="k">if</span> <span class="ow">not</span> <span class="n">counter_genes_compat_systems</span><span class="o">.</span><span class="n">has_key</span><span class="p">(</span><span class="n">syst</span><span class="o">.</span><span class="n">name</span><span class="p">):</span>
                            <span class="n">counter_genes_compat_systems</span><span class="p">[</span><span class="n">syst</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">counter_genes_compat_systems</span><span class="p">[</span><span class="n">syst</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">cur_compatible</span> <span class="o">=</span> <span class="n">compatible_systems</span>
            <span class="c1">#print [syst.name for syst in cur_compatible]            </span>
            <span class="n">cur_cluster</span> <span class="o">=</span> <span class="n">Cluster</span><span class="p">(</span><span class="n">cluster</span><span class="o">.</span><span class="n">systems_to_detect</span><span class="p">)</span> <span class="c1"># NEW</span>
            <span class="n">cur_cluster</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>

    <span class="n">cur_cluster</span><span class="o">.</span><span class="n">save</span><span class="p">()</span>         
   
    <span class="c1"># Check cluster status before storing it or not:</span>
    <span class="k">if</span> <span class="n">cur_cluster</span><span class="o">.</span><span class="n">state</span> <span class="o">==</span> <span class="s2">&quot;clear&quot;</span><span class="p">:</span>
        <span class="c1">#print &quot;\nclear to store: &quot;</span>
        <span class="c1">#print cur_cluster</span>
        <span class="n">res_clusters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_cluster</span><span class="p">)</span> 
        <span class="k">for</span> <span class="n">syst</span> <span class="ow">in</span> <span class="n">cur_compatible</span><span class="p">:</span> <span class="c1"># Good list of compatible??</span>
            <span class="c1">#print syst.name</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">counter_genes_compat_systems</span><span class="o">.</span><span class="n">has_key</span><span class="p">(</span><span class="n">syst</span><span class="o">.</span><span class="n">name</span><span class="p">):</span>
                <span class="n">counter_genes_compat_systems</span><span class="p">[</span><span class="n">syst</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cur_cluster</span><span class="o">.</span><span class="n">hits</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">counter_genes_compat_systems</span><span class="p">[</span><span class="n">syst</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cur_cluster</span><span class="o">.</span><span class="n">hits</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1">#print &quot;\nnope to store: &quot;</span>
        <span class="c1">#print cur_cluster</span>
        <span class="k">for</span> <span class="n">h_clust</span> <span class="ow">in</span> <span class="n">cur_cluster</span><span class="o">.</span><span class="n">hits</span><span class="p">:</span>
            <span class="c1">#h_compat = h_clust.gene.get_compatible_systems(cluster.systems_to_detect) # tmp before yep</span>
            <span class="n">h_compat</span> <span class="o">=</span> <span class="n">h_clust</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">get_compatible_systems</span><span class="p">(</span><span class="n">cluster</span><span class="o">.</span><span class="n">systems_to_detect</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># tmp before nope</span>
            <span class="k">for</span> <span class="n">syst</span> <span class="ow">in</span> <span class="n">h_compat</span><span class="p">:</span>
                <span class="c1">#print syst.name</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">counter_genes_compat_systems</span><span class="o">.</span><span class="n">has_key</span><span class="p">(</span><span class="n">syst</span><span class="o">.</span><span class="n">name</span><span class="p">):</span>
                    <span class="n">counter_genes_compat_systems</span><span class="p">[</span><span class="n">syst</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">counter_genes_compat_systems</span><span class="p">[</span><span class="n">syst</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="c1">#print &quot;\n---Counter compatible systems ! ---\n{0}&quot;.format(counter_genes_compat_systems)</span>

    <span class="c1"># Now final check: return only sub-clusters that consist in hits from systems represented at a single locus in the cluster to disambiguate.</span>
    <span class="n">real_res</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">res_clusters</span><span class="p">:</span>
        <span class="c1">#print &quot;\nres_cluster : &quot;</span>
        <span class="c1">#print r</span>
        <span class="n">nb_genes</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">hits</span><span class="p">)</span>
        <span class="c1">#print nb_genes</span>
        <span class="n">store</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">r</span><span class="o">.</span><span class="n">compatible_systems</span><span class="p">:</span>
            <span class="c1">#print c</span>
            <span class="k">if</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">counter_genes_compat_systems</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">counter_genes_compat_systems</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">!=</span> <span class="n">nb_genes</span><span class="p">:</span>
                    <span class="n">store</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">if</span> <span class="n">counter_genes_compat_systems</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">==</span> <span class="n">nb_genes</span><span class="p">:</span>
                    <span class="n">store</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">break</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">store</span> <span class="o">=</span> <span class="kc">False</span>                    
        <span class="k">if</span> <span class="n">store</span><span class="p">:</span>
            <span class="c1">#print &quot;=&gt; store !&quot;</span>
            <span class="n">real_res</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">real_res</span><span class="p">:</span>
        <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Store: &quot;</span><span class="p">)</span>
        <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>

    <span class="c1">#return res_clusters</span>
    <span class="k">return</span> <span class="n">real_res</span></div>


<div class="viewcode-block" id="analyze_clusters_replicon"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.analyze_clusters_replicon">[docs]</a><span class="k">def</span> <span class="nf">analyze_clusters_replicon</span><span class="p">(</span><span class="n">clusters</span><span class="p">,</span> <span class="n">systems</span><span class="p">,</span> <span class="n">multi_systems_genes</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Analyzes sets of contiguous hits (clusters) stored in a ClustersHandler for system detection:</span>

<span class="sd">    - split clusters if needed</span>
<span class="sd">    - delete them if they are not relevant</span>
<span class="sd">    - add eventual genes from other systems &quot;multi_system&quot; genes</span>
<span class="sd">    - check the QUORUM for each system to detect, *i.e.* mandatory + accessory - forbidden</span>

<span class="sd">    Only for \&quot;ordered\&quot; datasets representing a whole replicon.</span>
<span class="sd">    Reports systems occurence.</span>

<span class="sd">    :param clusters: the set of clusters to analyze</span>
<span class="sd">    :type clusters: :class:`macsypy.search_systems.ClustersHandler` </span>
<span class="sd">    :param systems: the set of systems to detect</span>
<span class="sd">    :type systems: a list of :class:`macsypy.system.System`</span>
<span class="sd">    :param multi_systems_genes: a dictionary with genes that could belong to multiple systems (keys are system names)</span>
<span class="sd">    :return: a set of systems occurence filled with hits found in clusters</span>
<span class="sd">    :rtype: a list of :class:`macsypy.search_systems.SystemOccurence` </span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Global Hits collectors, for uncomplete cluster Hits</span>
    <span class="n">systems_occurences_scattered</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">systems_occurences_list</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">syst_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">system</span> <span class="ow">in</span> <span class="n">systems</span><span class="p">:</span>
        <span class="n">syst_dict</span><span class="p">[</span><span class="n">system</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">system</span>
        <span class="n">systems_occurences_scattered</span><span class="p">[</span><span class="n">system</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">SystemOccurence</span><span class="p">(</span><span class="n">system</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">clust</span> <span class="ow">in</span> <span class="n">clusters</span><span class="o">.</span><span class="n">clusters</span><span class="p">:</span>
        <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">clust</span><span class="p">))</span>
        <span class="c1">#if clust.state == &quot;clear&quot;:</span>
        <span class="n">systems_to_consider</span> <span class="o">=</span> <span class="n">get_compatible_systems</span><span class="p">([</span><span class="n">system_bank</span><span class="p">[</span><span class="n">s</span><span class="p">]</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">clust</span><span class="o">.</span><span class="n">compatible_systems</span><span class="p">],</span> <span class="n">clust</span><span class="o">.</span><span class="n">systems_to_detect</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">clust</span><span class="o">.</span><span class="n">state</span> <span class="o">==</span> <span class="s2">&quot;clear&quot;</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">systems_to_consider</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># Local Hits collector</span>
            <span class="c1"># Check the putative system belongs to the list of systems to detect !! If it does not, do not go further with this cluster of genes.</span>
            <span class="c1"># New! different compatible systems are tested: then update cluster._putative_system w the good one?</span>
            <span class="c1">#print clust.compatible_systems</span>
            <span class="c1"># Arbitratily, if none of the set of compatible_systems pass the decision rule step, then the 1st system will store a scattered version of this...</span>
            <span class="n">first</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="c1">#store_scattered=True</span>
            <span class="n">store_scattered</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="n">store_clust</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="c1">#store_so = None # unused_var</span>
            <span class="c1">#store_msg = &quot;&quot; # unused_var</span>
            <span class="c1">#exclude = False # unused_var</span>
            <span class="c1">#for putative_system in clust.compatible_systems:</span>
            <span class="k">for</span> <span class="n">putative_system</span> <span class="ow">in</span> <span class="p">[</span><span class="n">s</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">systems_to_consider</span><span class="p">]:</span>
                <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Considering </span><span class="si">{}</span><span class="s2">: &quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">putative_system</span><span class="p">))</span>
                <span class="k">if</span> <span class="n">putative_system</span> <span class="ow">in</span> <span class="n">syst_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="n">so</span> <span class="o">=</span> <span class="n">SystemOccurence</span><span class="p">(</span><span class="n">syst_dict</span><span class="p">[</span><span class="n">putative_system</span><span class="p">])</span>
                    <span class="n">so</span><span class="o">.</span><span class="n">fill_with_cluster</span><span class="p">(</span><span class="n">clust</span><span class="p">)</span>
                    <span class="c1"># NEW!</span>
                    <span class="k">if</span> <span class="n">putative_system</span> <span class="ow">in</span> <span class="n">multi_systems_genes</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                        <span class="n">so</span><span class="o">.</span><span class="n">fill_with_multi_systems_genes</span><span class="p">(</span><span class="n">multi_systems_genes</span><span class="p">[</span><span class="n">putative_system</span><span class="p">])</span>
                    <span class="n">msg</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">decision_rule</span><span class="p">()</span>
                    <span class="n">so_state</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">state</span>
                    <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;-&gt; </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">so_state</span><span class="p">))</span>
                    <span class="k">if</span> <span class="n">so_state</span> <span class="o">!=</span> <span class="s2">&quot;exclude&quot;</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">so_state</span> <span class="o">!=</span> <span class="s2">&quot;single_locus&quot;</span><span class="p">:</span>
                            <span class="c1"># Store it to pool genes found with genes from other clusters.</span>
                            <span class="c1"># Do not do it if the so has a forbidden gene !!!</span>
                            <span class="c1"># NEW !! Now do not do this at the 1st try ! only if not complete is stored in the loop ! </span>
                            <span class="k">if</span> <span class="n">first</span><span class="p">:</span>
                                <span class="n">store_clust</span> <span class="o">=</span> <span class="n">clust</span>
                                <span class="n">store_scattered</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
                            <span class="n">systems_occurences_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">so</span><span class="p">)</span>
                            <span class="n">store_scattered</span> <span class="o">=</span> <span class="kc">False</span>
                            <span class="k">break</span>
            <span class="k">if</span> <span class="n">store_scattered</span><span class="p">:</span>
                <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;=&gt; Putative </span><span class="si">{}</span><span class="s2"> locus stored for later treatment of scattered systems.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">clust</span><span class="o">.</span><span class="n">compatible_systems</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
                <span class="n">systems_occurences_scattered</span><span class="p">[</span><span class="n">clust</span><span class="o">.</span><span class="n">compatible_systems</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">fill_with_cluster</span><span class="p">(</span><span class="n">store_clust</span><span class="p">)</span>

        <span class="k">elif</span> <span class="n">clust</span><span class="o">.</span><span class="n">state</span> <span class="o">==</span> <span class="s2">&quot;ambiguous&quot;</span><span class="p">:</span>
            <span class="c1"># Implement a way to &quot;clean&quot; the clusters. For instance :</span>
            <span class="c1"># - split the cluster in two if it seems that two systems are nearby</span>
            <span class="c1"># - remove single hits that are not forbidden for the &quot;main&quot; system and that are at one end of the current cluster</span>
            <span class="c1"># in this case, check that they are not &quot;loners&quot;, cause &quot;loners&quot; can be stored.</span>
            <span class="n">disamb_clusters</span> <span class="o">=</span> <span class="n">disambiguate_cluster</span><span class="p">(</span><span class="n">clust</span><span class="p">)</span>
            <span class="c1"># Add those new clusters to the set of clusters to fill system_occurrences?</span>
            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">disamb_clusters</span><span class="p">:</span>
                <span class="n">clusters</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">c</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">disamb_clusters</span><span class="p">:</span>
                <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;=&gt; disambiguated cluster(s) stored for later treatment&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;=&gt; none kept&quot;</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;------- next -------&quot;</span><span class="p">)</span>
    <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    </span>
<span class="s2">****************************************************</span>
<span class="s2">******* Report scattered/uncompleted systems *******</span>
<span class="s2">****************************************************</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">system</span> <span class="ow">in</span> <span class="n">systems</span><span class="p">:</span>
        <span class="c1">#print systems_occurences[system]</span>

        <span class="c1"># So new: Add support for the multi_loci parameter:</span>
        <span class="k">if</span> <span class="n">system</span><span class="o">.</span><span class="n">multi_loci</span><span class="p">:</span>

            <span class="n">so</span> <span class="o">=</span> <span class="n">systems_occurences_scattered</span><span class="p">[</span><span class="n">system</span><span class="o">.</span><span class="n">name</span><span class="p">]</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">decision_rule</span><span class="p">()</span>
            <span class="n">so_state</span> <span class="o">=</span> <span class="n">so</span><span class="o">.</span><span class="n">state</span>
            <span class="c1">#print &quot;-- {0} --&quot;.format(so.system_name)</span>
            <span class="c1">#print so</span>
            <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">so</span><span class="o">.</span><span class="n">is_complete</span><span class="p">():</span>
                <span class="n">systems_occurences_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">so</span><span class="p">)</span>
    <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;******************************************</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Stores results in this list? Or code a new object : systemDetectionReport ? </span>
    <span class="k">return</span> <span class="n">systems_occurences_list</span></div>


<div class="viewcode-block" id="build_clusters"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.build_clusters">[docs]</a><span class="k">def</span> <span class="nf">build_clusters</span><span class="p">(</span><span class="n">hits</span><span class="p">,</span> <span class="n">systems_to_detect</span><span class="p">,</span> <span class="n">rep_info</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Gets sets of contiguous hits according to the minimal inter_gene_max_space between two genes. Only for \&quot;ordered\&quot; datasets.</span>

<span class="sd">    :param hits: a list of Hmmer hits to analyze </span>
<span class="sd">    :type hits: a list of :class:`macsypy.report.Hit`</span>
<span class="sd">    :param systems_to_detect: the list of systems to detect </span>
<span class="sd">    :type systems_to_detect: a list of :class:`macsypy.system.System`</span>
<span class="sd">    :param cfg: the configuration object built from default and user parameters.</span>
<span class="sd">    :type cfg: :class:`macsypy.config.Config`</span>
<span class="sd">    :param rep_info: an entry extracted from the :class:`macsypy.database.RepliconDB`</span>
<span class="sd">    :type rep_info: a namedTuple &quot;RepliconInfo&quot; :class:`macsypy.database.RepliconInfo`</span>
<span class="sd">    :return: a set of clusters and a dictionary with \&quot;multi_system\&quot; genes stored in a system-wise way for further utilization.</span>
<span class="sd">    :rtype: :class:`macsypy.search_systems.ClustersHandler`</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">_log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Starting cluster detection with build_clusters... &quot;</span><span class="p">)</span>

    <span class="c1"># Deals with different dataset types using Pipeline ?? </span>
    <span class="n">clusters</span> <span class="o">=</span> <span class="n">ClustersHandler</span><span class="p">()</span>
    <span class="n">prev</span> <span class="o">=</span> <span class="n">hits</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1">#cur_cluster = Cluster()</span>
    <span class="n">cur_cluster</span> <span class="o">=</span> <span class="n">Cluster</span><span class="p">(</span><span class="n">systems_to_detect</span><span class="p">)</span>
    <span class="n">positions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">loner_state</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="c1"># New: storage of multi_system genes:</span>
    <span class="n">multi_system_genes_system_wise</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Before the case where there is a single hit was not treated...</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    if len(hits)==1 and prev.gene.loner:</span>
<span class="sd">        #print &quot;LONELY HITS LONER...&quot;</span>
<span class="sd">        if positions.count(prev.position) == 0:</span>
<span class="sd">            cur_cluster.add(prev)</span>
<span class="sd">            #clusters.add(cur_cluster)</span>
<span class="sd">            # New : Storage of multi_system genes:</span>
<span class="sd">            if prev.gene.multi_system:</span>
<span class="sd">                if not prev.system.name in multi_system_genes_system_wise.keys():</span>
<span class="sd">                    multi_system_genes_system_wise[prev.system.name] = []</span>
<span class="sd">                    multi_system_genes_system_wise[prev.system.name].append(prev)</span>

<span class="sd">            positions.append(prev.position)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">tmp</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">for</span> <span class="n">cur</span> <span class="ow">in</span> <span class="n">hits</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>

        <span class="n">_log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Hit </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cur</span><span class="p">))</span>
        <span class="c1">#prev_max_dist = prev.get_syst_inter_gene_max_space()</span>
        <span class="c1">#cur_max_dist = cur.get_syst_inter_gene_max_space()</span>
        <span class="n">prev_max_dist</span> <span class="o">=</span> <span class="n">prev</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">inter_gene_max_space</span>
        <span class="n">cur_max_dist</span> <span class="o">=</span> <span class="n">cur</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">inter_gene_max_space</span>
        <span class="n">inter_gene</span> <span class="o">=</span> <span class="n">cur</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span> <span class="o">-</span> <span class="n">prev</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span>

        <span class="n">tmp</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">****</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">tmp</span> <span class="o">+=</span> <span class="s2">&quot;prev_max_dist : </span><span class="si">{0:d}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">prev_max_dist</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">+=</span> <span class="s2">&quot;cur_max_dist : </span><span class="si">{0:d}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cur_max_dist</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">+=</span> <span class="s2">&quot;Intergene space : </span><span class="si">{0:d}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">inter_gene</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">+=</span> <span class="s2">&quot;Cur : </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cur</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">+=</span> <span class="s2">&quot;Prev : </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">prev</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">+=</span> <span class="s2">&quot;Len cluster: </span><span class="si">{0:d}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cur_cluster</span><span class="p">))</span>
        <span class="c1">#print tmp</span>

        <span class="c1"># First condition removes duplicates (hits for the same sequence)</span>
        <span class="c1"># the two others takes into account either system1 parameter or system2 parameter</span>

        <span class="c1">#smaller_dist = min(prev_max_dist, cur_max_dist)</span>
        <span class="k">if</span><span class="p">(</span><span class="n">inter_gene</span> <span class="o">&lt;=</span> <span class="n">prev_max_dist</span> <span class="ow">or</span> <span class="n">inter_gene</span> <span class="o">&lt;=</span> <span class="n">cur_max_dist</span><span class="p">):</span>
        <span class="c1">#if(inter_gene &lt;= smaller_dist ):</span>
            <span class="c1">#print &quot;zero&quot;</span>
            <span class="k">if</span> <span class="n">positions</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">prev</span><span class="o">.</span><span class="n">position</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1">#print &quot;un - ADD prev in cur_cluster&quot;</span>
                <span class="n">cur_cluster</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">prev</span><span class="p">)</span>
                <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prev</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>
                <span class="c1"># New : Storage of multi_system genes:</span>
                <span class="k">if</span> <span class="n">prev</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">multi_system</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">prev</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">multi_system_genes_system_wise</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                        <span class="n">multi_system_genes_system_wise</span><span class="p">[</span><span class="n">prev</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="n">multi_system_genes_system_wise</span><span class="p">[</span><span class="n">prev</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">name</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prev</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">positions</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">cur</span><span class="o">.</span><span class="n">position</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1">#print &quot;deux - ADD cur in cur_cluster&quot;</span>
                <span class="n">cur_cluster</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">cur</span><span class="p">)</span>
                <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>
                <span class="c1"># New : Storage of multi_system genes:</span>
                <span class="k">if</span> <span class="n">cur</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">multi_system</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">cur</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">multi_system_genes_system_wise</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                        <span class="n">multi_system_genes_system_wise</span><span class="p">[</span><span class="n">cur</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="n">multi_system_genes_system_wise</span><span class="p">[</span><span class="n">cur</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">name</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">prev</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">loner</span><span class="p">:</span>
                <span class="c1">#print &quot;trois - loner_state&quot;</span>
                <span class="c1">#print &quot;--- PREVLONER {0} {1}&quot;.format(prev.id, prev.gene.name)</span>
                <span class="n">loner_state</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Storage of the previous cluster</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cur_cluster</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c1">#print cur_cluster</span>
                <span class="c1">#print &quot;quatre - ADD cur_cluster&quot;</span>
                <span class="n">clusters</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">cur_cluster</span><span class="p">)</span> <span class="c1"># Add an in-depth copy of the object? </span>
                <span class="c1">#print(cur_cluster)</span>

                <span class="c1">#cur_cluster = Cluster()</span>
                <span class="n">cur_cluster</span> <span class="o">=</span> <span class="n">Cluster</span><span class="p">(</span><span class="n">systems_to_detect</span><span class="p">)</span>
                <span class="n">loner_state</span> <span class="o">=</span> <span class="kc">False</span>

            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">cur_cluster</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">loner_state</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span> <span class="c1"># WTF?</span>
                <span class="c1">#print cur_cluster</span>
                <span class="c1">#print &quot;cinq - ADD cur_cluster&quot;</span>
                <span class="c1">#print &quot;PREVLONER {0} {1}&quot;.format(prev.id, prev.gene.name)</span>
                <span class="n">clusters</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">cur_cluster</span><span class="p">)</span> <span class="c1"># Add an in-depth copy of the object? </span>

                <span class="n">cur_cluster</span> <span class="o">=</span> <span class="n">Cluster</span><span class="p">(</span><span class="n">systems_to_detect</span><span class="p">)</span>
                <span class="n">loner_state</span> <span class="o">=</span> <span class="kc">False</span>

            <span class="k">if</span> <span class="n">prev</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">loner</span><span class="p">:</span>
                <span class="c1">#print &quot;six - check&quot;</span>
                <span class="c1">#print &quot;PREVLONER ?? {0} {1}&quot;.format(prev.id, prev.gene.name)</span>

                <span class="k">if</span> <span class="n">positions</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">prev</span><span class="o">.</span><span class="n">position</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="c1">#print &quot;six - ADD prev in cur_cluster, ADD cur_cluster&quot;</span>
                    <span class="n">cur_cluster</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">prev</span><span class="p">)</span>
                    <span class="n">clusters</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">cur_cluster</span><span class="p">)</span>
                    <span class="c1">#print(cur_cluster)</span>

                    <span class="c1"># New : Storage of multi_system genes:</span>
                    <span class="k">if</span> <span class="n">prev</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">multi_system</span><span class="p">:</span>
                        <span class="k">if</span> <span class="ow">not</span> <span class="n">prev</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">multi_system_genes_system_wise</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                            <span class="n">multi_system_genes_system_wise</span><span class="p">[</span><span class="n">prev</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
                        <span class="n">multi_system_genes_system_wise</span><span class="p">[</span><span class="n">prev</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">name</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prev</span><span class="p">)</span>

                    <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prev</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>
                    <span class="n">loner_state</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="n">cur_cluster</span> <span class="o">=</span> <span class="n">Cluster</span><span class="p">(</span><span class="n">systems_to_detect</span><span class="p">)</span>

            <span class="n">cur_cluster</span> <span class="o">=</span> <span class="n">Cluster</span><span class="p">(</span><span class="n">systems_to_detect</span><span class="p">)</span>

        <span class="n">prev</span> <span class="o">=</span> <span class="n">cur</span>
        <span class="c1">#print &quot;Now prev is {0}&quot;.format(prev.gene.name)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cur_cluster</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cur_cluster</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">prev</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">loner</span><span class="p">):</span>
        <span class="c1">#print &quot;Recap clusters&quot;</span>
        <span class="n">clusters</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">cur_cluster</span><span class="p">)</span>

    <span class="c1"># Deal both with the case of single loner hits, and of last hits that are loners... YES!</span>
    <span class="c1">#print len(cur_cluster)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cur_cluster</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">prev</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">loner</span><span class="p">:</span>
        <span class="c1">#print &quot;FIFO or LIFO?&quot;</span>
        <span class="n">cur_cluster</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">prev</span><span class="p">)</span>
        <span class="n">clusters</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">cur_cluster</span><span class="p">)</span>
        <span class="c1"># New : Storage of multi_system genes:</span>
        <span class="k">if</span> <span class="n">prev</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">multi_system</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">prev</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">multi_system_genes_system_wise</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">multi_system_genes_system_wise</span><span class="p">[</span><span class="n">prev</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">multi_system_genes_system_wise</span><span class="p">[</span><span class="n">prev</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">name</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prev</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">rep_info</span><span class="o">.</span><span class="n">topology</span> <span class="o">==</span> <span class="s2">&quot;circular&quot;</span><span class="p">:</span>
        <span class="c1"># Need to take into account the possibility of a single gene at both extremity, </span>
        <span class="c1"># that should be considered as part of a cluster (and was not with previous steps)! </span>
        <span class="n">first</span> <span class="o">=</span> <span class="n">hits</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">last</span> <span class="o">=</span> <span class="n">hits</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">hitstoconsider</span><span class="o">=</span><span class="p">[]</span>
        <span class="c1">#_log_out.info(&quot;\n*** Check if single hits at ends are to consider for circularization ***\n&quot;)</span>
        <span class="k">if</span> <span class="n">positions</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">first</span><span class="o">.</span><span class="n">position</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">hitstoconsider</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">first</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">positions</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">last</span><span class="o">.</span><span class="n">position</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">hitstoconsider</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">last</span><span class="p">)</span>

        <span class="c1">#for h in hitstoconsider:</span>
        <span class="c1">#    _log_out.info(h)</span>
        <span class="n">clusters</span><span class="o">.</span><span class="n">circularize</span><span class="p">(</span><span class="n">rep_info</span><span class="p">,</span> <span class="n">hitstoconsider</span><span class="p">,</span> <span class="n">systems_to_detect</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">clusters</span><span class="p">,</span> <span class="n">multi_system_genes_system_wise</span><span class="p">)</span></div>

<div class="viewcode-block" id="get_best_hits"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.get_best_hits">[docs]</a><span class="k">def</span> <span class="nf">get_best_hits</span><span class="p">(</span><span class="n">hits</span><span class="p">,</span> <span class="n">tosort</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">criterion</span><span class="o">=</span><span class="s2">&quot;score&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns from a putatively redundant list of hits, a list of best matching hits.</span>
<span class="sd">    Analyzes quorum and co-localization if required for system detection. </span>
<span class="sd">    By default, hits are already sorted by position, and the hit with the best score is kept, then the best i-evalue. Possible criteria are:</span>

<span class="sd">    - maximal score (criterion=\&quot;score\&quot;)</span>
<span class="sd">    - minimal i-evalue (criterion=\&quot;i_eval\&quot;)</span>
<span class="sd">    - maximal percentage of the profile covered by the alignment with the query sequence (criterion=\&quot;profile_coverage\&quot;)</span>

<span class="sd">    :param tosort: tells if the hits have to be sorted</span>
<span class="sd">    :type tosort: boolean</span>
<span class="sd">    :param criterion: the criterion to base the sorting on </span>
<span class="sd">    :type criterion: string</span>
<span class="sd">    :return: the list of best matching hits</span>
<span class="sd">    :rtype: list of :class:`macsypy.report.Hit`</span>
<span class="sd">    :raise: a :class:`macsypy.macsypy_error.MacsypyError`</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">tosort</span><span class="p">:</span>
        <span class="n">hits</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">hits</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="n">operator</span><span class="o">.</span><span class="n">attrgetter</span><span class="p">(</span><span class="s1">&#39;position&#39;</span><span class="p">))</span>
    <span class="n">best_hits</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">prev_hit</span> <span class="o">=</span> <span class="n">hits</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">prev_pos</span> <span class="o">=</span> <span class="n">prev_hit</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">hits</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
        <span class="n">pos</span> <span class="o">=</span> <span class="n">h</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">pos</span> <span class="o">!=</span> <span class="n">prev_pos</span><span class="p">:</span>
            <span class="n">best_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prev_hit</span><span class="p">)</span>
            <span class="c1">#print &quot;******* no comp ****&quot;</span>
            <span class="c1">#print prev_hit</span>
            <span class="c1">#print &quot;******* ****** ****&quot;</span>
            <span class="n">prev_hit</span> <span class="o">=</span> <span class="n">h</span>
            <span class="n">prev_pos</span> <span class="o">=</span> <span class="n">pos</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1">#print &quot;******* COMP ****&quot;</span>
            <span class="c1">#print h </span>
            <span class="c1">#print prev_hit</span>
            <span class="k">if</span> <span class="n">criterion</span> <span class="o">==</span> <span class="s2">&quot;score&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">prev_hit</span><span class="o">.</span><span class="n">score</span> <span class="o">&lt;</span> <span class="n">h</span><span class="o">.</span><span class="n">score</span><span class="p">:</span>
                    <span class="n">prev_hit</span> <span class="o">=</span> <span class="n">h</span>
                <span class="c1"># To be tested before adding it !</span>
                <span class="c1">#if prev_hit.score == h.score:</span>
                <span class="c1">#    print prev_hit</span>
                <span class="c1">#    print h</span>
                <span class="c1">#    prev_hit = get_best_hits([prev_hit,h], False, i_eval)[0]</span>
            <span class="k">elif</span> <span class="n">criterion</span> <span class="o">==</span> <span class="s2">&quot;i_eval&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">prev_hit</span><span class="p">,</span> <span class="s1">&#39;i_eval&#39;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">h</span><span class="p">,</span> <span class="s1">&#39;i_eval&#39;</span><span class="p">):</span>
                    <span class="n">prev_hit</span> <span class="o">=</span> <span class="n">h</span>
            <span class="k">elif</span> <span class="n">criterion</span> <span class="o">==</span> <span class="s2">&quot;profile_coverage&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">prev_hit</span><span class="p">,</span> <span class="s1">&#39;profile_coverage&#39;</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">h</span><span class="p">,</span> <span class="s1">&#39;profile_coverage&#39;</span><span class="p">):</span>
                    <span class="n">prev_hit</span> <span class="o">=</span> <span class="n">h</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">MacsypyError</span><span class="p">(</span><span class="s1">&#39;The criterion for Hits comparison </span><span class="si">{}</span><span class="s1"> does not exist or is not available. </span><span class="se">\n</span><span class="s1">It must be either &quot;score&quot;, &quot;i_eval&quot; or &quot;profile_coverage&quot;.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">criterion</span><span class="p">))</span>

            <span class="c1">#print &quot;BEST&quot;</span>
            <span class="c1">#print prev_hit</span>
            <span class="c1">#print &quot;******* ****** ****&quot;</span>

    <span class="n">best_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prev_hit</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">best_hits</span></div>


<div class="viewcode-block" id="search_systems"><a class="viewcode-back" href="../../search_systems.html#macsypy.search_systems.search_systems">[docs]</a><span class="k">def</span> <span class="nf">search_systems</span><span class="p">(</span><span class="n">hits</span><span class="p">,</span> <span class="n">systems</span><span class="p">,</span> <span class="n">cfg</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Runs search of systems from a set of hits. Criteria for system assessment will depend on the kind of input dataset provided: </span>

<span class="sd">      - analyze **quorum and co-localization** for &quot;ordered_replicon&quot; and &quot;gembase&quot; datasets.</span>
<span class="sd">      - analyze **quorum only** (and in a limited way) for &quot;unordered_replicon&quot; and &quot;unordered&quot; datasets.</span>

<span class="sd">    :param hits: the list of hits for input systems components</span>
<span class="sd">    :type hits: list of :class:`macsypy.report.Hit`</span>
<span class="sd">    :param systems: the list of systems asked for detection</span>
<span class="sd">    :type systems: list of :class:`macsypy.system.System`</span>
<span class="sd">    :param cfg: the configuration object</span>
<span class="sd">    :type cfg: :class:`macsypy.config.Config`</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">tabfilename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">working_dir</span><span class="p">,</span> <span class="s1">&#39;macsyfinder.tab&#39;</span><span class="p">)</span>
    <span class="n">reportfilename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">working_dir</span><span class="p">,</span> <span class="s1">&#39;macsyfinder.report&#39;</span><span class="p">)</span>
    <span class="n">summaryfilename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">working_dir</span><span class="p">,</span> <span class="s1">&#39;macsyfinder.summary&#39;</span><span class="p">)</span>


    <span class="c1"># For the headers of the output files: no report so far ! print them in the loop at the 1st round ! </span>
    <span class="c1"># Update to fit only to the states looked for:</span>
    <span class="c1">#system_occurences_states = [&#39;single_locus&#39;, &#39;multi_loci&#39;]</span>
    <span class="n">system_occurences_states</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;single_locus&#39;</span><span class="p">]</span>
    <span class="n">system_names</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">multi_loci</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">systems</span><span class="p">:</span>
        <span class="n">syst_name</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">name</span>
        <span class="n">system_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">syst_name</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">s</span><span class="o">.</span><span class="n">multi_loci</span><span class="p">:</span>
            <span class="n">multi_loci</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="n">multi_loci</span><span class="p">:</span>
        <span class="n">system_occurences_states</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;multi_loci&#39;</span><span class="p">)</span>

    <span class="c1"># Specify to build_clusters the rep_info (min, max positions,[gene_name,...), and replicon_type... </span>
    <span class="c1"># Test with psae_circular_test.prt: pos_min = 1 , pos_max = 5569</span>
    <span class="c1">#RepInfo= namedtuple(&#39;RepInfo&#39;, [&#39;topology&#39;, &#39;min&#39;, &#39;max&#39;])</span>
    <span class="c1">#rep_info=RepInfo(&quot;circular&quot;, 1, 5569)</span>

    <span class="n">header_print</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">if</span> <span class="n">cfg</span><span class="o">.</span><span class="n">db_type</span> <span class="o">==</span> <span class="s1">&#39;gembase&#39;</span><span class="p">:</span>
        <span class="c1"># Construction of the replicon database storing info on replicons: </span>
        <span class="n">rep_db</span> <span class="o">=</span> <span class="n">RepliconDB</span><span class="p">(</span><span class="n">cfg</span><span class="p">)</span>
        <span class="n">replicons_w_hits</span><span class="o">=</span><span class="p">[]</span>
        <span class="n">json_all_systems</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># Use of the groupby() function from itertools : allows to group Hits by replicon_name, </span>
        <span class="c1"># and then apply the same build_clusters functions to replicons from &quot;gembase&quot; and &quot;ordered_replicon&quot; types of databases.</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">hits</span><span class="p">,</span> <span class="n">operator</span><span class="o">.</span><span class="n">attrgetter</span><span class="p">(</span><span class="s1">&#39;replicon_name&#39;</span><span class="p">)):</span>
            <span class="n">sub_hits</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>
            <span class="n">rep_info</span> <span class="o">=</span> <span class="n">rep_db</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>

            <span class="c1"># The following applies to any &quot;replicon&quot;</span>
            <span class="p">(</span><span class="n">clusters</span><span class="p">,</span> <span class="n">multi_syst_genes</span><span class="p">)</span> <span class="o">=</span> <span class="n">build_clusters</span><span class="p">(</span><span class="n">sub_hits</span><span class="p">,</span> <span class="n">systems</span><span class="p">,</span> <span class="n">rep_info</span><span class="p">)</span>          
            <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">************************************</span><span class="se">\n</span><span class="s2"> Analyzing clusters for </span><span class="si">{0}</span><span class="s2"> </span><span class="se">\n</span><span class="s2">************************************&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">))</span>
            <span class="c1"># Make analyze_clusters_replicon return an object systemOccurenceReport?</span>
            <span class="c1"># Note: at this stage, ther is no control of which systems are looked for... But systemsOccurrence do not have to be created for systems not searched. </span>
            <span class="c1"># </span>
            <span class="n">systems_occurences_list</span> <span class="o">=</span> <span class="n">analyze_clusters_replicon</span><span class="p">(</span><span class="n">clusters</span><span class="p">,</span> <span class="n">systems</span><span class="p">,</span> <span class="n">multi_syst_genes</span><span class="p">)</span>  

            <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;******************************************&quot;</span><span class="p">)</span>
            <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Building reports for </span><span class="si">{0}</span><span class="s2">: </span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">))</span>
            <span class="n">report</span> <span class="o">=</span> <span class="n">systemDetectionReportOrdered</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">systems_occurences_list</span><span class="p">,</span> <span class="n">cfg</span><span class="p">)</span>

            <span class="c1"># TO DO: Add replicons with no hits in tabulated_output!!! But where?! No trace of these replicons as replicons are taken from hits. </span>
            <span class="n">report</span><span class="o">.</span><span class="n">tabulated_output</span><span class="p">(</span><span class="n">system_occurences_states</span><span class="p">,</span> <span class="n">system_names</span><span class="p">,</span> <span class="n">tabfilename</span><span class="p">,</span> <span class="n">header_print</span><span class="p">)</span>
            <span class="n">report</span><span class="o">.</span><span class="n">report_output</span><span class="p">(</span><span class="n">reportfilename</span><span class="p">,</span> <span class="n">header_print</span><span class="p">)</span>
            <span class="n">report</span><span class="o">.</span><span class="n">summary_output</span><span class="p">(</span><span class="n">summaryfilename</span><span class="p">,</span> <span class="n">rep_info</span><span class="p">,</span> <span class="n">header_print</span><span class="p">)</span>

            <span class="n">json_all_systems</span> <span class="o">+=</span> <span class="n">report</span><span class="o">.</span><span class="n">system_2_json</span><span class="p">(</span><span class="n">rep_db</span><span class="p">)</span>
            <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;******************************************&quot;</span><span class="p">)</span>
            <span class="n">header_print</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="c1"># To add replicons with no systems in the </span>
            <span class="n">replicons_w_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>

        <span class="n">json_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">working_dir</span><span class="p">,</span> <span class="n">report</span><span class="o">.</span><span class="n">json_file_name</span><span class="p">)</span>
        <span class="n">report</span><span class="o">.</span><span class="n">json_output</span><span class="p">(</span><span class="n">json_path</span><span class="p">,</span> <span class="n">json_all_systems</span><span class="p">)</span>

        <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">--- Replicons with no hits: ---&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">tabfilename</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">_f</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">replicon</span> <span class="ow">in</span> <span class="n">rep_db</span><span class="o">.</span><span class="n">replicon_names</span><span class="p">():</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">replicon</span> <span class="ow">in</span> <span class="n">replicons_w_hits</span><span class="p">:</span>
                    <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">replicon</span><span class="p">)</span>
                    <span class="n">texte</span> <span class="o">=</span> <span class="n">replicon</span><span class="o">+</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">0&quot;</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">system_names</span><span class="p">)</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">system_occurences_states</span><span class="p">)</span><span class="o">+</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
                    <span class="c1">#print texte.strip()</span>
                    <span class="n">_f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">texte</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">cfg</span><span class="o">.</span><span class="n">db_type</span> <span class="o">==</span> <span class="s1">&#39;ordered_replicon&#39;</span><span class="p">:</span>
        <span class="c1"># Basically the same as for &#39;gembase&#39; (except the loop on replicons)</span>
        <span class="n">rep_db</span> <span class="o">=</span> <span class="n">RepliconDB</span><span class="p">(</span><span class="n">cfg</span><span class="p">)</span>
        <span class="n">rep_info</span> <span class="o">=</span> <span class="n">rep_db</span><span class="p">[</span><span class="n">RepliconDB</span><span class="o">.</span><span class="n">ordered_replicon_name</span><span class="p">]</span>

        <span class="c1">#(clusters, multi_syst_genes)=build_clusters(hits, rep_info) </span>
        <span class="p">(</span><span class="n">clusters</span><span class="p">,</span> <span class="n">multi_syst_genes</span><span class="p">)</span><span class="o">=</span><span class="n">build_clusters</span><span class="p">(</span><span class="n">hits</span><span class="p">,</span> <span class="n">systems</span><span class="p">,</span> <span class="n">rep_info</span><span class="p">)</span> 
        <span class="c1">#for syst in multi_syst_genes:</span>
        <span class="c1">#    for g in multi_syst_genes[syst]:</span>
        <span class="c1">#        print g</span>
        <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">************************************</span><span class="se">\n</span><span class="s2"> Analyzing clusters </span><span class="se">\n</span><span class="s2">************************************</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="c1">#systems_occurences_list = analyze_clusters_replicon(clusters, systems)</span>
        <span class="n">systems_occurences_list</span> <span class="o">=</span> <span class="n">analyze_clusters_replicon</span><span class="p">(</span><span class="n">clusters</span><span class="p">,</span> <span class="n">systems</span><span class="p">,</span> <span class="n">multi_syst_genes</span><span class="p">)</span>
        <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;******************************************&quot;</span><span class="p">)</span>
        <span class="c1">#print &quot;Reporting detected systems : \n&quot;</span>
        <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Building reports of detected systems</span><span class="se">\n</span><span class="s2"> &quot;</span><span class="p">)</span>
        <span class="n">report</span> <span class="o">=</span> <span class="n">systemDetectionReportOrdered</span><span class="p">(</span><span class="n">RepliconDB</span><span class="o">.</span><span class="n">ordered_replicon_name</span><span class="p">,</span> <span class="n">systems_occurences_list</span><span class="p">,</span> <span class="n">cfg</span><span class="p">)</span>
        <span class="n">report</span><span class="o">.</span><span class="n">tabulated_output</span><span class="p">(</span><span class="n">system_occurences_states</span><span class="p">,</span> <span class="n">system_names</span><span class="p">,</span> <span class="n">tabfilename</span><span class="p">,</span> <span class="n">header_print</span><span class="p">)</span>
        <span class="n">report</span><span class="o">.</span><span class="n">report_output</span><span class="p">(</span><span class="n">reportfilename</span><span class="p">,</span> <span class="n">header_print</span><span class="p">)</span>
        <span class="n">report</span><span class="o">.</span><span class="n">summary_output</span><span class="p">(</span><span class="n">summaryfilename</span><span class="p">,</span> <span class="n">rep_info</span><span class="p">,</span> <span class="n">header_print</span><span class="p">)</span>

        <span class="n">json_all_systems</span> <span class="o">=</span> <span class="n">report</span><span class="o">.</span><span class="n">system_2_json</span><span class="p">(</span><span class="n">rep_db</span><span class="p">)</span>
        <span class="n">json_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">working_dir</span><span class="p">,</span> <span class="n">report</span><span class="o">.</span><span class="n">json_file_name</span><span class="p">)</span>
        <span class="n">report</span><span class="o">.</span><span class="n">json_output</span><span class="p">(</span><span class="n">json_path</span><span class="p">,</span> <span class="n">json_all_systems</span><span class="p">)</span>
        <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;******************************************&quot;</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">cfg</span><span class="o">.</span><span class="n">db_type</span> <span class="o">==</span> <span class="s1">&#39;unordered_replicon&#39;</span> <span class="ow">or</span> <span class="n">cfg</span><span class="o">.</span><span class="n">db_type</span> <span class="o">==</span> <span class="s1">&#39;unordered&#39;</span><span class="p">:</span>

        <span class="c1"># implement a new function &quot;analyze_cluster&quot; =&gt; Fills a systemOccurence per system</span>
        <span class="n">systems_occurences_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># Hits with best score are first selected. </span>
        <span class="n">hits</span> <span class="o">=</span> <span class="n">get_best_hits</span><span class="p">(</span><span class="n">hits</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
        <span class="c1"># Then system-wise treatment:</span>
        <span class="n">hits</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">hits</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="n">operator</span><span class="o">.</span><span class="n">attrgetter</span><span class="p">(</span><span class="s1">&#39;system&#39;</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">hits</span><span class="p">,</span> <span class="n">operator</span><span class="o">.</span><span class="n">attrgetter</span><span class="p">(</span><span class="s1">&#39;system&#39;</span><span class="p">)):</span>
            <span class="c1"># SO new : if we want to include forbidden genes, </span>
            <span class="c1"># we have to get the corresponding list of hits at this point, </span>
            <span class="c1"># even if this is not their original system... </span>
            <span class="c1"># Need to compute the list of forbidden genes from hits for each system... </span>
            <span class="k">if</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">systems</span><span class="p">:</span>
                <span class="c1"># SO new: get the list of forbidden genes... Then have from hits </span>
                <span class="c1"># Should better rewrite this part of the code to have a single process of the hits...</span>
                <span class="c1">#forbidden_genes = k.forbidden_genes # unused_var</span>
                <span class="n">forbidden_hits</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">hits</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">h</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">is_forbidden</span><span class="p">(</span><span class="n">k</span><span class="p">):</span>
                        <span class="n">forbidden_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>
                <span class="n">sub_hits</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">g</span><span class="p">)</span> <span class="o">+</span> <span class="n">forbidden_hits</span>

                <span class="n">so</span> <span class="o">=</span> <span class="n">SystemOccurence</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
                <span class="c1">#resy=so.fill_with_hits(sub_hits) # does not return anything</span>
                <span class="c1">#so.fill_with_hits(sub_hits)</span>
                <span class="n">so</span><span class="o">.</span><span class="n">fill_with_hits</span><span class="p">(</span><span class="n">sub_hits</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span> <span class="c1"># SO new parameter to say wether forbidden genes should be included or not. </span>
                <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;******************************************&quot;</span><span class="p">)</span>
                <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;******************************************&quot;</span><span class="p">)</span>
                <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">so</span><span class="p">)</span>
                <span class="n">systems_occurences_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">so</span><span class="p">)</span>
        <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;******************************************&quot;</span><span class="p">)</span>
        <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Building reports of detected systems &quot;</span><span class="p">)</span>
        <span class="c1">#report = systemDetectionReportUnordered(systems_occurences_list, systems)</span>
        <span class="n">report</span> <span class="o">=</span> <span class="n">systemDetectionReportUnordered</span><span class="p">(</span><span class="n">systems_occurences_list</span><span class="p">,</span> <span class="n">cfg</span><span class="p">)</span>
        <span class="n">report</span><span class="o">.</span><span class="n">report_output</span><span class="p">(</span><span class="n">reportfilename</span><span class="p">,</span> <span class="n">header_print</span><span class="p">)</span>
        <span class="n">report</span><span class="o">.</span><span class="n">summary_output</span><span class="p">(</span><span class="n">summaryfilename</span><span class="p">,</span> <span class="n">header_print</span><span class="p">)</span>
        <span class="n">json_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cfg</span><span class="o">.</span><span class="n">working_dir</span><span class="p">,</span> <span class="n">report</span><span class="o">.</span><span class="n">json_file_name</span><span class="p">)</span>
        <span class="n">report</span><span class="o">.</span><span class="n">json_output</span><span class="p">(</span><span class="n">json_path</span><span class="p">)</span>
        <span class="n">_log_out</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;******************************************&quot;</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid database type. &quot;</span><span class="p">)</span></div>

</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">MacSyFinder documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../index.html" >Module code</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2014, Sophie Abby, Bertrand Néron.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.3.
    </div>
  </body>
</html>